{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identify procedural reasons for 'all-nothing-all' mutations (Part 1)\n",
    "## April 4, 2017\n",
    "\n",
    "Mutations that have 100% frequency in one generation, disappear completely (0%) in the next generation sampled, then reappear at 100% frequency. Potential explanations: \n",
    "\n",
    "1. Biological - Frequency-dependent selection? \n",
    "2. Procedural - Undersampled populations? Missing coverage during sequencing? Breseq frequency threshold for polymorphism mode (5-95%) not met?\n",
    "\n",
    "Input: \n",
    "1. HTML file from gdtools COMPARE\n",
    "2. GD files (annotated.gd) from breseq output.\n",
    "\n",
    "Tasks:\n",
    "1. Identify rows with 'all-nothing-all' (a-n-a) pattern in input HTML file, by mutation position.\n",
    "2. For each a-n-a, identify mutation position and generation(s) when 0% frequency occurs.\n",
    "3. For each 0% frequency occurrence in a-n-a, refer to relevant annotated.gd file to identify reason for frequency.\n",
    "4. Use prototype scripts here to build methods. See [lost_polymorphisms.py](https://github.com/yeemey/dvh_mms2/blob/master/lost_polymorphisms.py).\n",
    "\n",
    "#### Note:\n",
    "\n",
    "The following Part 1 prototype scripts have since been improved upon in [lost_polymorphisms.py](https://github.com/yeemey/dvh_mms2/blob/master/lost_polymorphisms.py) and no longer reflect what was used in analyses.\n",
    "\n",
    "```\n",
    "from bs4 import BeautifulSoup\n",
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "ua3 = BeautifulSoup(open('/Users/ymseah/Documents/compare/UA3.html'), 'html.parser')\n",
    "\n",
    "#Task 1-2\n",
    "\n",
    "#1. Identify line-generation headers\n",
    "line_generations = []\n",
    "next_line_gen = ua3.find('th', string=re.compile('0'))\n",
    "line_gen = next_line_gen.string\n",
    "while line_gen != 'annotation':\n",
    "    line_generations.append(line_gen)\n",
    "    next_line_gen = next_line_gen.next_element.next_element.next_element\n",
    "    line_gen = next_line_gen.string\n",
    "print(line_generations)\n",
    "\n",
    "#2. Identify ref genome and polymorphism position\n",
    "#3. Identify frequencies across generations\n",
    "#4. Link 2. and 3.\n",
    "\n",
    "table_cells = ua3.body.find_all('td')\n",
    "generation_frequencies_dict = {}\n",
    "for cell in table_cells:\n",
    "    if re.match('NC_', str(cell.string)):\n",
    "        position = cell.next_sibling.next_sibling.next_sibling\n",
    "        clean_position = int(re.sub(',', '', position.string))\n",
    "        gen_freqs_key = (cell.string, clean_position)\n",
    "        mutation = position.next_sibling.next_sibling.next_sibling\n",
    "        freq_ancestor = mutation.next_sibling.next_sibling.next_sibling\n",
    "        freq_gen100 = freq_ancestor.next_sibling\n",
    "        freq_gen300 = freq_gen100.next_sibling\n",
    "        freq_gen500 = freq_gen300.next_sibling\n",
    "        freq_gen780 = freq_gen500.next_sibling\n",
    "        freq_gen1000 = freq_gen780.next_sibling\n",
    "        gen_freqs_value = [mutation.string, freq_ancestor.string, freq_gen100.string, freq_gen300.string, freq_gen500.string, freq_gen780.string, freq_gen1000.string]\n",
    "        generation_frequencies_dict[gen_freqs_key] = gen_freqs_value\n",
    "\n",
    "#5. Identify a-n-a pattern.\n",
    "\n",
    "suspect_frequencies_dict = {}\n",
    "with open('suspect_frequencies.tsv', 'w') as output_file:\n",
    "    output_file.write('ref_genome\\tposition\\tmutation\\tfreq_anc\\tfreq_100\\tfreq_300\\tfreq_500\\tfreq_780\\tfreq_1000\\n')\n",
    "    for key, value in generation_frequencies_dict.items():\n",
    "        counter = 2\n",
    "        while counter <= len(value):\n",
    "            if counter + 1 < len(value):\n",
    "                if value[counter] == None:\n",
    "                    if value[counter - 1] == '100%' or value[counter + 1] == '100%':\n",
    "                        suspect_frequencies_dict[key] = value\n",
    "                        output_file.write(key[0] + '\\t' + str(key[1]) + '\\t')\n",
    "                        for item in value:\n",
    "                            output_file.write(str(item) + '\\t')\n",
    "                        output_file.write('\\n')\n",
    "                elif value[counter] == '100%':\n",
    "                    if value[counter - 1] == None or value[counter + 1] == None:\n",
    "                        suspect_frequencies_dict[key] = value\n",
    "                        output_file.write(key[0] + '\\t' + str(key[1]) + '\\t')\n",
    "                        for item in value:\n",
    "                            output_file.write(str(item) + '\\t')\n",
    "                        output_file.write('\\n')\n",
    "            counter += 2\n",
    "\n",
    "UA3_100 = pd.read_table('/Users/ymseah/Documents/sic_UA3-15_breseq/annotated.gd', comment='#', names=range(50), dtype=str)\n",
    "UA3_100.insert(0, 'generation', 100)\n",
    "UA3_300 = pd.read_table('/Users/ymseah/Documents/sic_UA3.45_breseq/annotated.gd', comment='#', names=range(50), dtype=str)\n",
    "UA3_300.insert(0, 'generation', 300)\n",
    "UA3_500 = pd.read_table('/Users/ymseah/Documents/sic_UA3-76_breseq/annotated.gd', comment='#', names=range(50), dtype=str)\n",
    "UA3_500.insert(0, 'generation', 500)\n",
    "UA3_780 = pd.read_table('/Users/ymseah/Documents/sic_UA3.118_breseq/annotated.gd', comment='#', names=range(50), dtype=str)\n",
    "UA3_780.insert(0, 'generation', 780)\n",
    "UA3_1000 = pd.read_table('/Users/ymseah/Documents/sic_UA3_S2_L001_breseq/output/evidence/annotated.gd', comment='#', names=range(50), dtype=str)\n",
    "UA3_1000.insert(0, 'generation', 1000)\n",
    "\n",
    "UA3_df  = pd.concat([UA3_100, UA3_300, UA3_500, UA3_780, UA3_1000], ignore_index=True)\n",
    "UA3_df.insert(0, 'line', 'UA3')\n",
    "UA3_df.insert(2, 'frequency', 0.0)\n",
    "UA3_df.insert(3, 'gene_product', '')\n",
    "UA3_df.insert(4, 'gene_position', '')\n",
    "UA3_df.insert(5, 'reject', '')\n",
    "\n",
    "for row in UA3_df.itertuples():\n",
    "    #check each column\n",
    "    col_index = 6\n",
    "    while col_index < 50:\n",
    "        #1. polymorphism frequencies\n",
    "        if re.match('frequency=', str(UA3_df.loc[row[0], col_index])):\n",
    "            UA3_df.loc[row[0], 'frequency'] = re.sub('frequency=', '', str(UA3_df.loc[row[0], col_index]))\n",
    "        #2. gene products\n",
    "        elif re.match('gene_product=', str(UA3_df.loc[row[0], col_index])):\n",
    "            UA3_df.loc[row[0], 'gene_product'] = re.sub('gene_product=', '', str(UA3_df.loc[row[0], col_index]))\n",
    "        #3. polymorphism rejection reasons\n",
    "        elif re.match('reject=', str(UA3_df.loc[row[0], col_index])):\n",
    "            UA3_df.loc[row[0], 'reject'] = re.sub('reject=', '', str(UA3_df.loc[row[0], col_index]))\n",
    "        #4. gene annotations\n",
    "        elif re.match('gene_position=', str(UA3_df.loc[row[0], col_index])):\n",
    "            UA3_df.loc[row[0], 'gene_position'] = re.sub('gene_position=', '', str(UA3_df.loc[row[0], col_index]))\n",
    "        col_index += 1\n",
    "    #set frequencies type to float\n",
    "    if re.match('1|2|3|4|5|6|7|8|9', str(UA3_df.loc[row[0], 'frequency'])):\n",
    "        UA3_df.loc[row[0], 'frequency'] = float(UA3_df.loc[row[0], 'frequency'])\n",
    "    else:\n",
    "        UA3_df.loc[row[0], 'frequency'] = 0.0\n",
    "    #set positions (col 4) type to int\n",
    "    UA3_df.loc[row[0], 4] = int(UA3_df.loc[row[0], 4])\n",
    "    #set reject col to 'NA' when no reject reason given.\n",
    "    if (UA3_df.loc[row[0], 'reject'] == '') & (UA3_df.loc[row[0], 2] == '.'):\n",
    "        UA3_df.loc[row[0], 'reject'] = 'NA'\n",
    "\n",
    "UA3_df.rename(columns = {0: 'entry_type', 1: 'item_id', 2: 'evidence_ids', 3: 'ref_genome', 4:'position'}, inplace=True)\n",
    "ua3df_subset = UA3_df[['line', 'generation', 'frequency', 'gene_product', 'gene_position', 'reject', 'entry_type', 'item_id', 'evidence_ids', 'ref_genome', 'position']].copy()\n",
    "ua3df_subset.to_csv('/Users/ymseah/Documents/ua3.csv', index=False)\n",
    "\n",
    "# Task 3.\n",
    "#6: Identify reasons for a-n-a frequencies.\n",
    "\n",
    "ua3df_subset_mutations_only = ua3df_subset[(ua3df_subset['entry_type'] == 'SNP') | \n",
    "                                           (ua3df_subset['entry_type'] == 'SUB') | \n",
    "                                           (ua3df_subset['entry_type'] == 'DEL') | \n",
    "                                           (ua3df_subset['entry_type'] == 'INS') | \n",
    "                                           (ua3df_subset['entry_type'] == 'MOB') | \n",
    "                                           (ua3df_subset['entry_type'] == 'AMP') | \n",
    "                                           (ua3df_subset['entry_type'] == 'CON') | \n",
    "                                           (ua3df_subset['entry_type'] == 'INV')]\n",
    "for key, value in suspect_frequencies_dict.items():\n",
    "    row_indices = ua3df_subset_mutations_only[(ua3df_subset_mutations_only['ref_genome'] == key[0]) & \n",
    "                                              (ua3df_subset_mutations_only['position'] == key[1])].index.tolist()\n",
    "    #print(key, row_indices)\n",
    "    for row in row_indices:\n",
    "        reject_reason = ua3df_subset_mutations_only.loc[row, 'reject']\n",
    "        generation = ua3df_subset_mutations_only.loc[row, 'generation']\n",
    "        entry_type = ua3df_subset_mutations_only.loc[row, 'entry_type']\n",
    "        #print(row, reject_reason, generation, entry_type)\n",
    "        if reject_reason != '':\n",
    "            if generation == 100:\n",
    "                value[2] = str(value[2]) + ' ' + reject_reason\n",
    "            elif generation == 300:\n",
    "                value[3] = str(value[3]) + ' ' + reject_reason\n",
    "            elif generation == 500:\n",
    "                value[4] = str(value[4]) + ' ' + reject_reason\n",
    "            elif generation == 780:\n",
    "                value[5] = str(value[5]) + ' ' + reject_reason\n",
    "            elif generation == 1000:\n",
    "                value[6] = str(value[6]) + ' ' + reject_reason\n",
    "```\n",
    "\n",
    "## Part 1 Problems & Solutions\n",
    "\n",
    "#### String sorting messes up generation order. \n",
    "\n",
    "```\n",
    "find_line_generations = set(ua3.find_all('th', string=re.compile('UA3')))\n",
    "line_generations=['0']\n",
    "for each in find_line_generations:\n",
    "    line_generations.append(each.string)\n",
    "    line_generations.sort()\n",
    "print(line_generations)\n",
    "```\n",
    "\n",
    "- Solved by appending in order of appearance in document.\n",
    "\n",
    "\n",
    "#### Identifying frequencies in each row, when there are no string elements within each cell to represent 0% frequency. \n",
    "\n",
    "\n",
    "```\n",
    "frequencies_by_generations = {}\n",
    "all_comments = ua3.find_all(text = lambda text:isinstance(text, Comment))\n",
    "for comment in all_comments:\n",
    "    if comment == ' Seq_Id ':\n",
    "        ref_genome = comment.previous_element.strip()\n",
    "        position = comment.next_element.next_element.next_element.strip()\n",
    "        genome_position_key = (ref_genome, position)\n",
    "        generation_frequencies_value = []\n",
    "        #gen_freq = comment.next_sibling.next_sibling.next_sibling.next_sibling.next_sibling.next_sibling.next_sibling.next_sibling.next_element.strip()\n",
    "        print(genome_position_key)\n",
    "```\n",
    "\n",
    "- Solved by listing each cell (td) as an element and searching within td. Initially, searching was by html comment position, but it got messy.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identify procedural reasons for 'all-nothing-all' mutations (Part 2)\n",
    "## April 24, 2017\n",
    "\n",
    "Tasks:\n",
    "1. Use output from [lost_polymorphisms_runScript.py](https://github.com/yeemey/dvh_mms2/blob/master/lost_polymorphisms_runScript.py) to classify procedural reasons, summary counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "ua3_freqs = pd.read_csv('~/Documents/compare/parse_compare/UA3_suspect_frequencies.tsv', sep='\\t')\n",
    "ua3_evidence = pd.read_csv('~/Documents/compare/parse_compare/UA3_rejected_evidence.tsv', sep='\\t')\n",
    "ua3freqs_evidence = pd.merge(ua3_freqs, ua3_evidence, how='left', on='position')\n",
    "\n",
    "ue3_freqs = pd.read_csv('~/Documents/compare/parse_compare/UE3_suspect_frequencies.tsv', sep='\\t')\n",
    "ue3_evidence = pd.read_csv('~/Documents/compare/parse_compare/UE3_rejected_evidence.tsv', sep='\\t')\n",
    "ue3freqs_evidence = pd.merge(ue3_freqs, ue3_evidence, how='left', on='position')\n",
    "\n",
    "ur1_freqs = pd.read_csv('~/Documents/compare/parse_compare/UR1_suspect_frequencies.tsv', sep='\\t')\n",
    "ur1_evidence = pd.read_csv('~/Documents/compare/parse_compare/UR1_rejected_evidence.tsv', sep='\\t')\n",
    "ur1freqs_evidence = pd.merge(ur1_freqs, ur1_evidence, how='left', on='position')\n",
    "\n",
    "us1_freqs = pd.read_csv('~/Documents/compare/parse_compare/US1_suspect_frequencies.tsv', sep='\\t')\n",
    "us1_evidence = pd.read_csv('~/Documents/compare/parse_compare/US1_rejected_evidence.tsv', sep='\\t')\n",
    "us1freqs_evidence = pd.merge(us1_freqs, us1_evidence, how='left', on='position')\n",
    "\n",
    "ha3_freqs = pd.read_csv('~/Documents/compare/parse_compare/HA3_suspect_frequencies.tsv', sep='\\t')\n",
    "ha3_evidence = pd.read_csv('~/Documents/compare/parse_compare/HA3_rejected_evidence.tsv', sep='\\t')\n",
    "ha3freqs_evidence = pd.merge(ha3_freqs, ha3_evidence, how='left', on='position')\n",
    "\n",
    "he3_freqs = pd.read_csv('~/Documents/compare/parse_compare/HE3_suspect_frequencies.tsv', sep='\\t')\n",
    "he3_evidence = pd.read_csv('~/Documents/compare/parse_compare/HE3_rejected_evidence.tsv', sep='\\t')\n",
    "he3freqs_evidence = pd.merge(he3_freqs, he3_evidence, how='left', on='position')\n",
    "\n",
    "hr2_freqs = pd.read_csv('~/Documents/compare/parse_compare/HR2_suspect_frequencies.tsv', sep='\\t')\n",
    "hr2_evidence = pd.read_csv('~/Documents/compare/parse_compare/HR2_rejected_evidence.tsv', sep='\\t')\n",
    "hr2freqs_evidence = pd.merge(hr2_freqs, hr2_evidence, how='left', on='position')\n",
    "\n",
    "hs3_freqs = pd.read_csv('~/Documents/compare/parse_compare/HS3_suspect_frequencies.tsv', sep='\\t')\n",
    "hs3_evidence = pd.read_csv('~/Documents/compare/parse_compare/HS3_rejected_evidence.tsv', sep='\\t')\n",
    "hs3freqs_evidence = pd.merge(hs3_freqs, hs3_evidence, how='left', on='position')\n",
    "\n",
    "#All results with suspect frequencies, i.e., None --> 100%, or 100% --> None\n",
    "#From COMPARE html output: ref_genome_x, position, mutation_x, html_freq__\n",
    "#From annotated.gd: ref_genome_y, mutation_y, line, generation, frequency, reject_reason, evidence_type, item_id\n",
    "all_freqs_evidence = pd.concat([ua3freqs_evidence, ue3freqs_evidence, ur1freqs_evidence, us1freqs_evidence, ha3freqs_evidence, he3freqs_evidence, hr2freqs_evidence, hs3freqs_evidence], ignore_index=True)\n",
    "all_freqs_evidence.to_csv('~/Documents/compare/parse_compare/all_freqs_evidence.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(251, 9)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_freqs = pd.concat([ua3_freqs, ue3_freqs, ur1_freqs, us1_freqs, ha3_freqs, he3_freqs, hr2_freqs, hs3_freqs], ignore_index=True)\n",
    "all_freqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x11079d9e8>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAFsCAYAAADLxNJsAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XlcVOXiBvBnZhARkWUYxMZ9Q8Gy9KIkmBujedXMStHK\nSi3NoMxuLumvq5VLlBJFuX1KKe1aWq7Xul4bK03IDcQNg3AnFGSRRcAB5vz+8DrXuQNiMyMHzvt8\n/5qzzMwDHB4O7zlzjkqSJAlERCQMtdwBiIiobrH4iYgEw+InIhIMi5+ISDAsfiIiwbD4iYgEw+In\nIhIMi5+ISDAsfiIiwbD4iYgE4yJ3gJpkZWXJHUERdDodcnNz5Y5BVC1un86j1+vveF3u8RMRCYbF\nT0QkGBY/EZFgWPxERIJh8RMRCYbFT0QkGBY/EZFgWPxERIJh8RMRCabefnKXiOxXNXmk3BHuSLbc\nAe6Q5tPtckdwKu7xExEJptY9/uXLlyM5ORleXl6IiYkBAMTGxlqupVNaWgp3d3csWbLE5rlRUVFw\nc3ODWq2GRqNBdHS0k+MTEdGfVWvxDxgwAEOHDsWyZcss81577TXL47Vr18Ld3b3G58+fPx+enp4O\nxiQiImepdagnKCgIHh4e1S6TJAm//vorwsLCnB6MiIjuDocO7p46dQpeXl645557alxnwYIFUKvV\nGDx4MAwGQ43rGY1GGI1GAEB0dDR0Op0j0eg/XFxc+L0UUEM5aNpQKO13yKHiT0hIuO3e/oIFC6DV\nalFYWIiFCxdCr9cjKCio2nUNBoPVHwZeo9s5eL1zIsc1hN+hOrkef1VVFQ4ePIjQ0NAa19FqtQAA\nLy8v9OrVCxkZGfa+HREROYndxX/8+HHo9Xr4+vpWu7y8vBxlZWWWx8eOHUObNm3sfTsiInKSWod6\nPvzwQ6SmpqK4uBhTp05FREQEBg0aVO0wT35+PlatWoU5c+agsLAQS5cuBXDjv4O+ffvigQceuDtf\nBRER3TGVJEmS3CGqw3vuOgfH+MXUUD6521A0hE/u8p67RERUIxY/EZFgWPxERIJh8RMRCYbFT0Qk\nGBY/EZFgWPxERIJh8RMRCYbFT0QkGBY/EZFgWPxERIJh8RMRCYbFT0QkGBY/EZFgWPxERIJh8RMR\nCYbFT0QkGBY/EZFgWPxERIJh8RMRCcalthWWL1+O5ORkeHl5ISYmBgCwceNG7N69G56engCAJ598\nEj179rR5bkpKCuLj42E2mxEeHo5Ro0Y5OT4REf1ZtRb/gAEDMHToUCxbtsxq/vDhwzFy5Mgan2c2\nm7F69Wq8+eab8PX1xZw5cxAcHIxWrVo5npqIiOxW61BPUFAQPDw8/vQLZ2RkoEWLFvD394eLiwtC\nQ0Nx6NAhu0ISEZHz1LrHX5OdO3di79696NChA5599lmbPw75+fnw9fW1TPv6+uL333+v8fWMRiOM\nRiMAIDo6Gjqdzt5odAsXFxd+LwWULXcAhVHa75BdxT9kyBCMHj0aALBhwwasXbsWkZGRDgUxGAww\nGAyW6dzcXIdej27Q6XT8XhI5qCH8Dun1+jte166zery9vaFWq6FWqxEeHo7Tp0/brKPVapGXl2eZ\nzsvLg1arteftiIjIiewq/oKCAsvjgwcPonXr1jbrdOzYEZcuXUJOTg4qKyuRmJiI4OBg+5MSEZFT\n1DrU8+GHHyI1NRXFxcWYOnUqIiIicPLkSZw7dw4qlQp+fn6YMmUKgBvj+qtWrcKcOXOg0WgwadIk\nLFq0CGazGQMHDqz2DwQREdUtlSRJktwhqpOVlSV3BEXgGL+YqibXfKo1/XmaT7fLHaFWd32Mn4iI\nGi4WPxGRYFj8RESCYfETEQmGxU9EJBgWPxGRYFj8RESCYfETEQmGxU9EJBgWPxGRYFj8RESCYfET\nEQmGxU9EJBgWPxGRYFj8RESCYfETEQmGxU9EJBgWPxGRYFj8RESCqfVm68uXL0dycjK8vLwQExMD\nAFi3bh2SkpLg4uICf39/REZGomnTpjbPjYqKgpubG9RqNTQaDaKjo53/FRAR0Z9Sa/EPGDAAQ4cO\nxbJlyyzzunfvjqeeegoajQZffvkltmzZgvHjx1f7/Pnz58PT09N5iYmIyCG1DvUEBQXBw8PDat79\n998PjUYDAAgICEB+fv7dSUdERE5X6x5/bX788UeEhobWuHzBggVQq9UYPHgwDAaDo29HREQOcqj4\nN2/eDI1Gg4ceeqja5QsWLIBWq0VhYSEWLlwIvV6PoKCgatc1Go0wGo0AgOjoaOh0Okei0X+4uLjw\neymgbLkDKIzSfofsLv6ff/4ZSUlJmDdvHlQqVbXraLVaAICXlxd69eqFjIyMGovfYDBY/UeQm5tr\nbzS6hU6n4/eSyEEN4XdIr9ff8bp2nc6ZkpKCbdu2Yfbs2WjcuHG165SXl6OsrMzy+NixY2jTpo09\nb0dERE5U6x7/hx9+iNTUVBQXF2Pq1KmIiIjAli1bUFlZiQULFgAAOnfujClTpiA/Px+rVq3CnDlz\nUFhYiKVLlwIAqqqq0LdvXzzwwAN396shIqJaqSRJkuQOUZ2srCy5IygCh3rEVDV5pNwRFEXz6Xa5\nI9Tqrg/1EBFRw8XiJyISDIufiEgwLH4iIsGw+ImIBMPiJyISDIufiEgwLH4iIsGw+ImIBMPiJyIS\nDIufiEgwLH4iIsE4fAcuUTWUi2A1lBtyNISLYBEpBff4iYgEw+InIhIMi5+ISDAsfiIiwbD4iYgE\nw+InIhIMi5+ISDC1nse/fPlyJCcnw8vLCzExMQCAkpISxMbG4sqVK/Dz88Nrr70GDw8Pm+empKQg\nPj4eZrMZ4eHhGDVqlPO/AiIi+lNq3eMfMGAA5s6dazVv69atuO+++xAXF4f77rsPW7dutXme2WzG\n6tWrMXfuXMTGxiIhIQGZmZnOS05ERHaptfiDgoJs9uYPHTqE/v37AwD69++PQ4cO2TwvIyMDLVq0\ngL+/P1xcXBAaGlrtekREVLfsGuMvLCyEj48PAMDb2xuFhYU26+Tn58PX19cy7evri/z8fDtjEhGR\nszh8rR6VSgWVSuVwEKPRCKPRCACIjo6GTqdz+DXvpoZyDZyGor7/vBsabp/OpbTt067i9/LyQkFB\nAXx8fFBQUABPT0+bdbRaLfLy8izTeXl50Gq1Nb6mwWCAwWCwTOfm5toTjRoo/rypPmsI26der7/j\nde0a6gkODsaePXsAAHv27EGvXr1s1unYsSMuXbqEnJwcVFZWIjExEcHBwfa8HREROVGte/wffvgh\nUlNTUVxcjKlTpyIiIgKjRo1CbGwsfvzxR8vpnMCNcf1Vq1Zhzpw50Gg0mDRpEhYtWgSz2YyBAwei\ndevWd/0LIiKi21NJkiTJHaI6WVlZcke4rYZyPf6Ggtfjdy5un87VELbPuz7UQ0REDReLn4hIMCx+\nIiLBsPiJiATD4iciEgyLn4hIMCx+IiLBsPiJiATD4iciEgyLn4hIMCx+IiLBsPiJiATD4iciEgyL\nn4hIMCx+IiLBsPiJiATD4iciEgyLn4hIMCx+IiLBsPiJiATjYu8Ts7KyEBsba5nOyclBREQEhg8f\nbpl38uRJvP/++2jevDkAICQkBKNHj3YgLhEROcru4tfr9ViyZAkAwGw248UXX0Tv3r1t1gsMDMQb\nb7xhf0IiInIqpwz1HD9+HC1atICfn58zXo6IiO4iu/f4b5WQkICwsLBql6WlpWHGjBnQarV45pln\n0Lp162rXMxqNMBqNAIDo6GjodDpnRLtrsuUOoDD1/efd0HD7dC6lbZ8OF39lZSWSkpLw1FNP2Sxr\n3749VqxYATc3NyQnJ2PJkiWIi4ur9nUMBgMMBoNlOjc319Fo1IDw5031WUPYPvV6/R2v6/BQz5Ej\nR9C+fXt4e3vbLHN3d4ebmxsAoGfPnqiqqkJRUZGjb0lERA5wuPhvN8xz9epVSJIEAMjIyIDZbEaz\nZs0cfUsiInKAQ0M95eXlOHbsGKZMmWKZt2vXLgDAkCFDsH//fuzatQsajQaurq6YPn06VCqVY4mJ\niMghKunmLnk9k5WVJXeE26qaPFLuCIqi+XS73BEUhdunczWE7bNOx/iJiKhhYfETEQmGxU9EJBgW\nPxGRYFj8RESCYfETEQmGxU9EJBgWPxGRYFj8RESCYfETEQmGxU9EJBgWPxGRYFj8RESCYfETEQmG\nxU9EJBgWPxGRYFj8RESCYfETEQmGxU9EJBgWPxGRYFwceXJUVBTc3NygVquh0WgQHR1ttVySJMTH\nx+PIkSNo3LgxIiMj0aFDB4cCExGRYxwqfgCYP38+PD09q1125MgRXL58GXFxcfj999/x2WefYfHi\nxY6+JREROeCuDvUcPnwY/fr1g0qlQkBAAK5du4aCgoK7+ZZERFQLh/f4FyxYALVajcGDB8NgMFgt\ny8/Ph06ns0z7+voiPz8fPj4+Nq9jNBphNBoBANHR0VbPq4+y5Q6gMPX9593QcPt0LqVtnw4V/4IF\nC6DValFYWIiFCxdCr9cjKCjIrtcyGAxWfzhyc3MdiUYNDH/eVJ81hO1Tr9ff8boODfVotVoAgJeX\nF3r16oWMjAyb5bd+w/Ly8izPISIiedhd/OXl5SgrK7M8PnbsGNq0aWO1TnBwMPbu3QtJkpCeng53\nd/dqh3mIiKju2D3UU1hYiKVLlwIAqqqq0LdvXzzwwAPYtWsXAGDIkCHo0aMHkpOTMW3aNLi6uiIy\nMtI5qYmIyG4qSZIkuUNUJysrS+4It1U1eaTcERRF8+l2uSMoCrdP52oI22edjfETEVHDw+InIhIM\ni5+ISDAsfiIiwbD4iYgEw+InIhIMi5+ISDAsfiIiwbD4iYgEw+InIhIMi5+ISDAsfiIiwbD4iYgE\nw+InIhIMi5+ISDAsfiIiwbD4iYgEw+InIhIMi5+ISDB232w9NzcXy5Ytw9WrV6FSqWAwGDBs2DCr\ndU6ePIn3338fzZs3BwCEhIRg9OjRjiUmIiKH2F38Go0GzzzzDDp06ICysjK88cYb6N69O1q1amW1\nXmBgIN544w2HgxIRkXPYPdTj4+ODDh06AACaNGmCli1bIj8/32nBiIjo7rB7j/9WOTk5OHv2LDp1\n6mSzLC0tDTNmzIBWq8UzzzyD1q1bO+MtiYjITg4Xf3l5OWJiYjBhwgS4u7tbLWvfvj1WrFgBNzc3\nJCcnY8mSJYiLi6v2dYxGI4xGIwAgOjoaOp3O0Wh3VbbcARSmvv+8Gxpun86ltO3ToeKvrKxETEwM\nHnroIYSEhNgsv/UPQc+ePbF69WoUFRXB09PTZl2DwQCDwWCZzs3NdSQaNTD8eVN91hC2T71ef8fr\n2j3GL0kSVq5ciZYtW2LEiBHVrnP16lVIkgQAyMjIgNlsRrNmzex9SyIicgK79/jT0tKwd+9etGnT\nBjNnzgQAPPnkk5a/jEOGDMH+/fuxa9cuaDQauLq6Yvr06VCpVM5JTkREdrG7+Lt27YqNGzfedp2h\nQ4di6NCh9r4FERHdBfzkLhGRYFj8RESCYfETEQmGxU9EJBgWPxGRYFj8RESCYfETEQmGxU9EJBgW\nPxGRYFj8RESCYfETEQmGxU9EJBgWPxGRYFj8RESCYfETEQmGxU9EJBgWPxGRYFj8RESCYfETEQmG\nxU9EJBi7b7YOACkpKYiPj4fZbEZ4eDhGjRpltVySJMTHx+PIkSNo3LgxIiMj0aFDB4cCExGRY+ze\n4zebzVi9ejXmzp2L2NhYJCQkIDMz02qdI0eO4PLly4iLi8OUKVPw2WefORyYiIgcY3fxZ2RkoEWL\nFvD394eLiwtCQ0Nx6NAhq3UOHz6Mfv36QaVSISAgANeuXUNBQYHDoYmIyH52F39+fj58fX0t076+\nvsjPz7dZR6fT3XYdIiKqWw6N8TuT0WiE0WgEAERHR0Ov18ucqBbfHZY7AVHNuH3Sbdi9x6/VapGX\nl2eZzsvLg1artVknNzf3tuvcZDAYEB0djejoaHsjUTXeeOMNuSMQ1YjbpzzsLv6OHTvi0qVLyMnJ\nQWVlJRITExEcHGy1TnBwMPbu3QtJkpCeng53d3f4+Pg4HJqIiOxn91CPRqPBpEmTsGjRIpjNZgwc\nOBCtW7fGrl27AABDhgxBjx49kJycjGnTpsHV1RWRkZFOC05ERPZRSZIkyR2C7h6j0QiDwSB3DKJq\ncfuUB4ufiEgwvGQDEZFgWPxERIJh8RMRCYbFrzA7d+60PL548aKMSYislZaW1rjs1s/70N3H4leY\nn376yfL4k08+kTEJkbW3337b8vidd96xWrZkyZK6jiM0Fr+C8YQtqk9u3R5LSkpqXEZ3X725Vg85\nx7Vr13Dw4EGYzWaUlZXhwIEDVstDQkJkSkaiU6lU1T6ubpruLha/wgQGBuLw4cOWx0lJSVbLWfwk\nl8LCQuzYsQOSJFkeAzf29ouKimROJxZ+gIuI6sQ333xz2+VjxoypoyTE4leYZcuWISoqCgDw888/\nY8CAAfIGIqJ6h0M9CnP+/HnL43/9618sfqo3Ll68iOzsbMtVfD///HPLKZ5Dhw7l/bjrEM/qURge\nJKP66h//+AeaNWtmmT569Ch69uyJe++9F99++62MycTDPX6FycvLw5o1a2we3zRp0iQ5YhHh6tWr\n6NKli2Xa3d0dDz74IADghx9+kCuWkFj8CjN+/HjLY/7rTPVJWVmZ1fSiRYssj3lWT91i8SvMrWP6\n5eXlAAA3NzeZ0hD9l1arxe+//47OnTtbzU9PT+ed+eoYz+pRoF27dmHLli24fv06gBvF/+ijj+Lh\nhx+WORmJLCMjA7GxsRgwYADat28PADhz5gz27NmD1157DZ06dZI5oThY/AqzadMmpKenY9KkSfD3\n9wcAZGdnIz4+Hp07d8YTTzwhc0ISWWFhIXbu3Gm5gGDr1q3x8MMPw9vbW+ZkYmHxK8yrr76KJUuW\nwNXV1Wq+yWTCzJkz8dFHH8mUjKh6ubm5SExMxMiRI+WOIgyO8SuMSqWyKX0AcHV15ameVG8UFRXh\n119/RUJCAgoKCtCrVy+5IwmFxa8wWq0Wx48fx3333Wc1/8SJEzyARrK6edHAhIQEZGVlISQkBDk5\nOVi5cqXc0YTD4leYiRMn4v3330fXrl0tp3OePn0aaWlpmDVrlszpSGQvvPACOnXqhHHjxqFr165Q\nqVQ4ePCg3LGExDF+BTKZTNi3bx8yMzMBAK1atULfvn2rHQIiqivfffcdEhMTUV5ejrCwMISGhmLh\nwoW8YZAMWPwKs3DhQrz55ptyxyCqUXZ2NhISEpCQkIDLly9jzJgx6N27N/R6vdzRhMFr9SgMPwFJ\n9Z2/vz8ef/xxxMTE4N1330VpaSneffdduWMJhWP8ClNaWmpz161b8UYsVJ+0adMG48aNQ6tWreSO\nIhQWv8KUlpba3HXrVix+kktpaSn+/e9/Iz8/H8HBwejevTt27tyJHTt2oG3btujXr5/cEYXB4lcY\nPz8/REZGyh2DyMYnn3yCpk2bIiAgALt378aWLVsgSRJmzpyJdu3ayR1PKCx+heGxeqqvsrOzERMT\nAwAIDw/HlClTsHz5cp5tJgMe3FWYV155Re4IRNVycfnvfqZarYavry9LXyY8nVNhnn32WZtLM3h6\neqJbt254+umnre6ARFSXxo4da7lEuCRJMJlMaNy4MSRJgkqlwhdffCFzQnGw+AVQUlKCn3/+Genp\n6fjb3/4mdxwikhmHegTg4eGBESNGIDs7W+4oJLATJ05YHufk5Fgtu90pyOR8LH5BVFZWoqqqSu4Y\nJLB169ZZHt88yHvT5s2b6zqO0HhWj8JUt+d07do1JCYmWm5sTSSHW0eV/3eEmSPOdYvFrzDVfXir\nWbNmGDZsGHr27ClDIqIbbj3p4H9PQOC9IuoWD+4SUZ2YMGECAgMDIUkSfvvtNwQGBgK4sbeflpaG\n+Ph4mROKg8WvMOvWrUOLFi0wePBgq/k//PADcnJy8PTTT8uUjESXmpp62+VBQUF1lIQ41KMwJ06c\nwPjx423mh4eHY+bMmSx+kk1GRgZCQ0Oh0+nkjiI8Fr/CVFZWVjteqlareQCNZJWfn4+///3v8PPz\nQ1hYGPr06QNPT0+5YwmJxa8wrq6uuHTpEu655x6r+ZcuXeLH40lWEyZMwHPPPYdTp04hISEBmzZt\nQtu2bREWFoaQkBA0adJE7ojC4Bi/whw5cgRr1qzBE088YXXP3a1bt+K5557jmT1Ub5jNZhw7dgzr\n169HVlYWvvzyS7kjCYPFr0AXLlzA9u3bcfHiRQBA69atMXLkSLRp00bmZEQ3XLhwAQkJCUhMTISn\npyfCwsIwbNgwuWMJg8UvqDVr1mDSpElyxyCBXLp0yVL2arUaoaGhCAsLg7+/v9zRhMMxfkGlpaXJ\nHYEEs3jxYoSFhWH69On871NmLH4iqhMff/wxgBsXaLv5CfNWrVpxj18GLH4iqhNlZWVYsWIFzp49\ni7Zt2wIAzp07hw4dOmDq1Klwd3eXOaE4WPyC4qEdqmtr1qxBq1atMH36dKjVNy4MLEkSNm3ahDVr\n1uDll1+WOaE4eFlmhUlMTLyj9XgGBdW1tLQ0REREWEofuHFxttGjRyM9PV3GZOJh8SvM3r17sWjR\nolpvujJgwIC6CUR0B/gfaN3SvPXWW2/JHYKcp2/fvnBzc8PKlStRUlKCVq1aoaKiAiaTCSaTiZ/e\nJdmcPn0aFy5cQGBgoNVlRb799ls0btwYISEhMqYTC8/jV6hz585h/vz58PDwsMxTqVT45JNPZExF\nIistLcXKlSutDu6eP38e7dq1w9SpU9G0aVOZE4qDxa8wFRUV2LRpEw4cOIDx48fjL3/5i9yRiKxc\nvnwZmZmZAG6cztmiRQuZE4mHxa8wr776KkJCQjB69GgO61C9kpKSgvLycptbgO7fvx/u7u7o3r27\nTMnEw4O7CjNz5kyMGDHCpvSLiopgMplkSkUEbNq0qdqbrQQFBWHDhg0yJBIXi19hvvvuO5w6dcpm\n/m+//YYvvvhChkREN1RUVFR7/X1PT09cv35dhkTiYvErzJkzZ6o9O6J3797V/kEgqitlZWWoqqqy\nmV9ZWcn/RusYi19hbvcLxMM5JKfevXtj1apVKC8vt8wrLy/Hp59+it69e8uYTDy8ZIPCeHp6IiMj\nA506dbKan5GRwdvckazGjRuHr7/+GlFRUZb77ubm5mLQoEEYO3aszOnEwrN6FCYjIwOxsbHo37+/\n5Q5cZ86cwZ49ezB9+nR07txZ5oQkOpPJhMuXLwMAWrRoYXMiwrFjx3iGz13G4legq1evYteuXbhw\n4QKAG3fgGjp0KLy8vGRORlS72bNn47333pM7hqJxqEeBvL29ERERIXcMIrtwX/TuY/ErzNtvv13j\nMpVKhXnz5tVhGqI/79br+NDdweJXmGeeecZmXnp6OrZv386hHiICwOJXnJsHdAEgNTUVmzZtgslk\nwuTJk9GjRw8ZkxHdGT8/P7kjKB4P7ipQSkoKNm/ejEaNGuGxxx7DvffeK3ckIgBAcXEx9u3bhz/+\n+APAjYu0hYWFoVmzZjInEwuLX2HmzJmDoqIiPPLIIwgICLBZfut/BER1KTMzE++88w7uv/9+tG/f\nHpIk4ezZszh+/DjmzZuHli1byh1RGCx+hXnrrbdue3Bs/vz5dZiG6L9iYmLQp08fhIaGWs3fv38/\n9u3bhxkzZsiUTDwc41cY3lCN6qsLFy7g9ddft5n/4IMP4quvvpIhkbh4rR6F2bZtm+Xxr7/+arVs\n/fr1dR2HyMLNzc2uZeR83ONXmMTERDz66KMAgK1bt6JPnz6WZUePHsVTTz0lVzQSXGFhIXbs2GEz\nX5IkFBUVyZBIXCx+hbn1kM3/Hr7h4RySU3h4OMrKyqpdNmjQoDpOIzYWv8LcemD3fw/y8hORJKcx\nY8bIHYH+g2f1KMzYsWPh5uYGSZJgMpnQuHFjADf29isqKngQjWTz7bff3nb56NGj6ygJsfiJqE78\n85//tJl3/fp1/PjjjyguLsa6detkSCUmFr/ClJSU2Mxr2rQph3moXikrK8P333+PH3/8EX369MEj\njzzCa0nVIRa/wkRFRUGlUlkdyC0vL0e7du3w4osvonnz5jKmI9GVlJRgx44d+OWXX9C/f38MGzYM\nHh4ecscSDotfEAcOHIDRaMT//d//yR2FBLVu3TocPHgQ4eHhGDp0KM/dlxGLXyC8sxHJaezYsXBx\ncYFGo7EaepQkCSqVCl988YWM6cTC0zkFUV5eDrPZLHcMEtiGDRvkjkD/weJXmOo+GVlSUoKkpCQ8\n/PDDMiQiqll5eTkOHjyIhIQEzJkzR+44wmDxK8z/fjJSpVLB29sbr7zyCtq0aSNTKqL/qqysRHJy\nMvbt24ejR48iJCQEgwcPljuWUDjGr2ClpaUAAHd3d5mTEN24VlRCQgKOHj2Kbt26ITQ0FPHx8Vi2\nbJnc0YTDPX4F+v7777Ft2zZUVFQAAJo1a4aIiAiEhYUhNzcXOp1O5oQkosWLF6Nr165YsGCB5bTi\nzz//XN5QgmLxK8zGjRtx+vRpvPPOO/D39wcAZGdnIz4+HleuXMHu3bvx8ccfy5ySRPTee+8hISHB\nUvxhYWE84UAmHOpRmGnTpmHp0qVwdXW1mm8ymfD888/j1VdfRXBwsEzpiG5IS0tDQkICDhw4gLZt\n26J3794wGAxyxxIG9/gVRq1W25Q+ALi6ukKr1bL0qV7o0qULunTpggkTJuD48eNITExk8dch3oFL\nYbRaLY4fP24z/8SJE9BqtTIkIqqZWq2Gn58fryVVx7jHrzATJ07E+++/j65du6JDhw4AgNOnTyMt\nLQ2zZ8+WOR2J7Pz581i3bh0KCgrQq1cvPPzww1i9ejUyMjIwYsQIueMJhWP8CmQymbBv3z5kZmYC\nAFq1aoW+fftWOwREVFfmzp2LIUOGICAgACkpKdiyZQv69++PiIgIbpt1jHv8CvPHH3+gZcuWGDRo\nECoqKtCoUSPLsvT0dAQEBMiYjkRWUVGBAQMGAAD0ej2+//57jB8/Xt5QguIYv8LExcVZHr/55ptW\ny1avXl0vB94+AAAIaklEQVTXcYgsKioqcPbsWZw5cwZnzpxBo0aNrKap7nCPX2F4s3Wqr3x8fLB2\n7VrLtLe3t9X0/Pnz5YglJBa/wvBm61RfPfnkkxxqrCdY/AqTl5eHNWvW2DwGgPz8fLliEWH16tW8\nH0Q9weJXmFsPlt08nbOmaaK6xKHG+oOncwqEF2gjOU2YMAGBgYE1LufnTOoO9/gVKD09Hfn5+QgM\nDISXlxfOnz+PrVu34rfffsOKFSvkjkeC8vT0xCOPPCJ3DAKLX3HWrVuH5ORktG3bFtu2bcP999+P\n3bt347HHHsNLL70kdzwSmJubG4KCguSOQWDxK05ycjLee+89uLq6oqSkBC+99BJiYmIs1z8nkkvT\npk1x9epVeHt7AwD27NmDAwcOQKfTISIiAh4eHjInFAc/wKUwrq6ulo+/e3h44J577mHpU71QWloK\nF5cb+5qpqalYv349+vXrB3d3d6xatUrmdGLhHr/CZGdnW50yl5OTYzXNA2gkF7PZbNmrT0xMRHh4\nOB588EE8+OCDmDlzpszpxMLiV5hZs2ZZTfNgGtUXZrMZVVVV0Gg0OHHiBKZMmWK1jOoOi19hmjdv\nzlM2qV4KCwvDW2+9hWbNmsHV1dVyaufly5fh7u4uczqx8Dx+hZk9e7ZlaGfp0qWYMWOGzImI/is9\nPR1Xr15F9+7d4ebmBgDIyspCeXk5P2BYh7jHrzC3/h3PycmRMQmRrequ1aPX62VIIjae1aMwt7tI\nGxERwKEexRk7dizc3NwgSRJMJhMaN24M4MZ/AiqVCl988YXMCYlIbix+IiLBcKiHiEgwLH4iIsGw\n+KnB2rx5M1auXCl3DKIGh2P8JLSTJ0/i448/5h8QEgr3+KneqqqqkjvCbdX3fEQ14R4/1StRUVEY\nPHgw9u3bh6ysLMTFxeHzzz/HqVOn4ObmhuHDh2PYsGEAgI0bN+Ly5cuYNm0agBufCl27di0yMzPh\n5+eHCRMmoFu3bgCAkpISrF27FkePHoXJZEJgYCCmTZuG559/HpWVlZYrmn700UfQarXVZtu4cSMu\nXryIRo0aISkpCc8++ywGDhyI7du3Y/fu3bh27RruvfdeTJkyxXIxsg8++ACnTp2CyWRCu3bt8MIL\nL6B169YAblxCe926dcjLy0OTJk0wfPhwjBw5EgBgNBqxbds2lJSUoGvXrpg8ebIlV0REBF544QXs\n2LEDRUVF6Nu3L55//nl+boPunERUj0RGRkozZsyQrly5IpWVlUmzZs2SvvnmG6miokK6fPmyFBUV\nJR05ckSSJEnasGGD9NFHH0mSJEl5eXnSxIkTpaSkJKmqqko6evSoNHHiRKmwsFCSJElavHix9MEH\nH0jFxcVSRUWFdPLkSUmSJOnEiRPSiy++eEfZNmzYII0bN046cOCAVFVVJV2/fl367rvvpLlz50q5\nubmSyWSSVq1aJcXGxlqes3v3bqm0tFQymUxSfHy8NGPGDMuyyZMnS6mpqZIkSVJxcbF0+vRpSZIk\n6fjx49KkSZOk06dPSyaTSVq9erU0b948y/PGjBkjvfvuu1JJSYl05coVadKkSZbvCdGd4FAP1Tt/\n/etfodPpcPHiRRQVFWH06NFwcXGBv78/wsPDkZiYaPOcvXv3okePHujZsyfUajW6d++Ojh07Ijk5\nGQUFBUhJScHkyZPh4eEBFxcXu+8EFRAQgN69e0OtVsPV1RU//PADxo0bB19fXzRq1AhjxozBgQMH\nLMNAgwYNQpMmTSzLzp8/j9LSUgCARqNBZmYmSktL4eHhYblWzS+//IKBAweiQ4cOaNSoEZ566imk\np6dbXYJj1KhRaNq0KXQ6Hbp164Zz587Z9fWQmHitHqp3bl5d9MqVKygoKMCECRMsy8xmc7U37M7N\nzcX+/fuRlJRkmVdVVYVu3bohLy8PHh4eTrnDk6+vr9X0lStXsHTpUqthFrVajcLCQnh7e+Orr77C\n/v37UVRUZFmnqKgI7u7ueP3117F582asX78ebdq0wdNPP42AgAAUFBSgffv2ltdzc3ODh4cH8vPz\nLTfVuXkXKwBo3LgxysvLHf7aSBwsfqq3dDodmjdvjri4uFrX9fX1xUMPPYSpU6faLCsoKEBJSQmu\nXbuGpk2bWi1zdFzc19cXL730Erp27WqzbO/evTh8+DD+/ve/w8/PD6WlpZg4caJleadOnTBr1ixU\nVlZi586diI2NxYoVK+Dj44Pc3FzLeuXl5SgpKanx2APRn8WhHqq3OnXqhCZNmmDr1q0wmUwwm824\ncOECMjIybNZ96KGHkJSUhJSUFJjNZphMJpw8eRJ5eXnw8fHBAw88gM8++wwlJSWorKxEamoqAMDL\nywvFxcWW4Zc/a/Dgwfj6669x5coVADf25g8dOgQAKCsrg4uLCzw8PHD9+nV89dVXludVVlbil19+\nsdyO0N3d3fJHKCwsDD/99BPOnTuHiooKfPXVV+jUqRNvoUlOwz1+qrfUajVmz56NtWvXIioqCpWV\nldDr9Rg7dqzNujqdDrNmzcKXX36Jjz76CGq1Gp06dcLkyZMBAK+88go+//xzvPbaa6isrES3bt0Q\nFBSEli1bIiwsDC+//DLMZjM++OCDP7VnffMMo4ULF6KgoABeXl7o06cPevXqhf79++Po0aOYOnUq\nPDw8MHbsWOzatcvy3L1792LNmjUwm83Q6/WWs5O6d++OsWPHIiYmBiUlJejSpQumT5/uyLeSyApP\n56QGa8OGDcjLy0NkZKTcUYgaFA71UIMkSRIyMzM5/EFkBw71UIM0e/ZsNGrUCM8//7xTX3fx4sU4\ndeqUzfzHHnsMjz/+uFPfi0guHOohIhIMh3qIiATD4iciEgyLn4hIMCx+IiLBsPiJiATD4iciEsz/\nA3zc34Wm0MttAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1107cb320>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib\n",
    "%matplotlib inline\n",
    "matplotlib.style.use('ggplot')\n",
    "\n",
    "#How many 'None' frequencies are due to:\n",
    "#1a. rejection based on evidence filters?\n",
    "rejected_mutation_reasons = all_freqs_evidence.groupby('reject_reason')['mutation_x'].count()\n",
    "rejected_mutation_reasons.plot.bar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "reject_reason\n",
       "FREQUENCY_CUTOFF    11\n",
       "STRAND_COVERAGE     18\n",
       "Name: mutation_x, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rejected_mutation_reasons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_genome_x</th>\n",
       "      <th>position</th>\n",
       "      <th>mutation_x</th>\n",
       "      <th>html_freq_anc</th>\n",
       "      <th>html_freq_100</th>\n",
       "      <th>html_freq_300</th>\n",
       "      <th>html_freq_500</th>\n",
       "      <th>html_freq_780</th>\n",
       "      <th>html_freq_1000</th>\n",
       "      <th>ref_genome_y</th>\n",
       "      <th>mutation_y</th>\n",
       "      <th>line</th>\n",
       "      <th>generation</th>\n",
       "      <th>frequency</th>\n",
       "      <th>polymorphism_frequency</th>\n",
       "      <th>new_cov</th>\n",
       "      <th>total_cov</th>\n",
       "      <th>reject_reason</th>\n",
       "      <th>evidence_type</th>\n",
       "      <th>item_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>708564</td>\n",
       "      <td>C→T</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>C→T</td>\n",
       "      <td>UA3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.818228</td>\n",
       "      <td>4/5</td>\n",
       "      <td>4/7</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>111.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>1376015</td>\n",
       "      <td>C→T</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>C→T</td>\n",
       "      <td>UA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.046179</td>\n",
       "      <td>0.046179</td>\n",
       "      <td>9/3</td>\n",
       "      <td>126/133</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>481.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3140594</td>\n",
       "      <td>+GA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+GA</td>\n",
       "      <td>UA3</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.800007</td>\n",
       "      <td>0.800007</td>\n",
       "      <td>20/24</td>\n",
       "      <td>29/26</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>267.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3140594</td>\n",
       "      <td>+GA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+GA</td>\n",
       "      <td>UA3</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.800014</td>\n",
       "      <td>0.800014</td>\n",
       "      <td>20/24</td>\n",
       "      <td>29/26</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>268.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3140594</td>\n",
       "      <td>+GA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+GA</td>\n",
       "      <td>UA3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.600002</td>\n",
       "      <td>0.600002</td>\n",
       "      <td>2/1</td>\n",
       "      <td>2/3</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>136.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3140594</td>\n",
       "      <td>+GA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+GA</td>\n",
       "      <td>UA3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.600006</td>\n",
       "      <td>0.600006</td>\n",
       "      <td>2/1</td>\n",
       "      <td>2/3</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>137.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>NC_005791</td>\n",
       "      <td>540893</td>\n",
       "      <td>A→G</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005791</td>\n",
       "      <td>A→G</td>\n",
       "      <td>UA3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.894805</td>\n",
       "      <td>0.894805</td>\n",
       "      <td>28/23</td>\n",
       "      <td>31/26</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>279.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>42867</td>\n",
       "      <td>+G</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+G</td>\n",
       "      <td>UR1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.926836</td>\n",
       "      <td>0.926836</td>\n",
       "      <td>17/21</td>\n",
       "      <td>18/23</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>227.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2083305</td>\n",
       "      <td>+C</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+C</td>\n",
       "      <td>UR1</td>\n",
       "      <td>500.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.934467</td>\n",
       "      <td>6/8</td>\n",
       "      <td>6/9</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>174.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2982788</td>\n",
       "      <td>+CC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+CC</td>\n",
       "      <td>UR1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.588240</td>\n",
       "      <td>0.588240</td>\n",
       "      <td>2/8</td>\n",
       "      <td>6/11</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>344.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2982788</td>\n",
       "      <td>+CC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+CC</td>\n",
       "      <td>UR1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.588240</td>\n",
       "      <td>0.588240</td>\n",
       "      <td>2/8</td>\n",
       "      <td>6/11</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>345.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2982788</td>\n",
       "      <td>+CC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+CC</td>\n",
       "      <td>UR1</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.500004</td>\n",
       "      <td>0.500004</td>\n",
       "      <td>1/2</td>\n",
       "      <td>3/3</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>186.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2982788</td>\n",
       "      <td>+CC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+CC</td>\n",
       "      <td>UR1</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.500004</td>\n",
       "      <td>0.500004</td>\n",
       "      <td>1/2</td>\n",
       "      <td>3/3</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>187.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>1144620</td>\n",
       "      <td>+C</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+C</td>\n",
       "      <td>US1</td>\n",
       "      <td>780.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.923082</td>\n",
       "      <td>5/7</td>\n",
       "      <td>5/8</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>212.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3140577</td>\n",
       "      <td>Δ2 bp</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>Δ2 bp</td>\n",
       "      <td>US1</td>\n",
       "      <td>500.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.937500</td>\n",
       "      <td>6/9</td>\n",
       "      <td>6/10</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>167.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3140594</td>\n",
       "      <td>+GA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+GA</td>\n",
       "      <td>US1</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.823488</td>\n",
       "      <td>0.823488</td>\n",
       "      <td>5/8</td>\n",
       "      <td>8/9</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>169.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3140594</td>\n",
       "      <td>+GA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+GA</td>\n",
       "      <td>US1</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.764713</td>\n",
       "      <td>0.764713</td>\n",
       "      <td>5/8</td>\n",
       "      <td>8/9</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>170.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3140594</td>\n",
       "      <td>+GA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+GA</td>\n",
       "      <td>US1</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.800004</td>\n",
       "      <td>0.800004</td>\n",
       "      <td>10/10</td>\n",
       "      <td>13/12</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>446.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3140594</td>\n",
       "      <td>+GA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+GA</td>\n",
       "      <td>US1</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.800004</td>\n",
       "      <td>0.800004</td>\n",
       "      <td>10/10</td>\n",
       "      <td>13/12</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>447.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>NC_005791</td>\n",
       "      <td>1216664</td>\n",
       "      <td>G→A</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005791</td>\n",
       "      <td>G→A</td>\n",
       "      <td>US1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.882416</td>\n",
       "      <td>4/11</td>\n",
       "      <td>5/12</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>1263.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>NC_005791</td>\n",
       "      <td>1648281</td>\n",
       "      <td>G→A</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005791</td>\n",
       "      <td>G→A</td>\n",
       "      <td>US1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.894879</td>\n",
       "      <td>0.894879</td>\n",
       "      <td>17/17</td>\n",
       "      <td>17/21</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>1326.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>30836</td>\n",
       "      <td>+A</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>+A</td>\n",
       "      <td>US1</td>\n",
       "      <td>500.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.857157</td>\n",
       "      <td>3/3</td>\n",
       "      <td>4/3</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>8236.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>1376015</td>\n",
       "      <td>C→T</td>\n",
       "      <td>None</td>\n",
       "      <td>?</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>C→T</td>\n",
       "      <td>HA3</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.942985</td>\n",
       "      <td>16/17</td>\n",
       "      <td>17/18</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2305774</td>\n",
       "      <td>Δ2 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>?</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>Δ2 bp</td>\n",
       "      <td>HA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.069762</td>\n",
       "      <td>0.069762</td>\n",
       "      <td>5/4</td>\n",
       "      <td>78/51</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>497.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2497231</td>\n",
       "      <td>+TC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+TC</td>\n",
       "      <td>HA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.039215</td>\n",
       "      <td>0.039215</td>\n",
       "      <td>3/3</td>\n",
       "      <td>86/67</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>522.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2497231</td>\n",
       "      <td>+TC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+TC</td>\n",
       "      <td>HA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.039215</td>\n",
       "      <td>0.039215</td>\n",
       "      <td>3/3</td>\n",
       "      <td>86/67</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>523.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3140594</td>\n",
       "      <td>+GA</td>\n",
       "      <td>None</td>\n",
       "      <td>?</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+GA</td>\n",
       "      <td>HA3</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.687505</td>\n",
       "      <td>0.687505</td>\n",
       "      <td>11/11</td>\n",
       "      <td>14/18</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>386.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3140594</td>\n",
       "      <td>+GA</td>\n",
       "      <td>None</td>\n",
       "      <td>?</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+GA</td>\n",
       "      <td>HA3</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.687508</td>\n",
       "      <td>0.687508</td>\n",
       "      <td>11/11</td>\n",
       "      <td>14/18</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "      <td>387.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2081507</td>\n",
       "      <td>Δ1 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>Δ1 bp</td>\n",
       "      <td>HR2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.033324</td>\n",
       "      <td>0.033324</td>\n",
       "      <td>4/1</td>\n",
       "      <td>77/73</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>327.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ref_genome_x  position mutation_x html_freq_anc html_freq_100  \\\n",
       "5      NC_002937    708564        C→T          None          None   \n",
       "17     NC_002937   1376015        C→T          None          None   \n",
       "28     NC_002937   3140594        +GA          None          None   \n",
       "29     NC_002937   3140594        +GA          None          None   \n",
       "32     NC_002937   3140594        +GA          None          None   \n",
       "33     NC_002937   3140594        +GA          None          None   \n",
       "46     NC_005791    540893        A→G          None          None   \n",
       "162    NC_002937     42867         +G          100%          None   \n",
       "179    NC_002937   2083305         +C          100%          100%   \n",
       "190    NC_002937   2982788        +CC          None          None   \n",
       "191    NC_002937   2982788        +CC          None          None   \n",
       "196    NC_002937   2982788        +CC          None          None   \n",
       "197    NC_002937   2982788        +CC          None          None   \n",
       "271    NC_002937   1144620         +C          100%          100%   \n",
       "291    NC_002937   3140577      Δ2 bp          100%          100%   \n",
       "300    NC_002937   3140594        +GA          None          None   \n",
       "301    NC_002937   3140594        +GA          None          None   \n",
       "302    NC_002937   3140594        +GA          None          None   \n",
       "303    NC_002937   3140594        +GA          None          None   \n",
       "325    NC_005791   1216664        G→A          None          None   \n",
       "337    NC_005791   1648281        G→A          None          None   \n",
       "354    NC_005863     30836         +A          100%          100%   \n",
       "386    NC_002937   1376015        C→T          None             ?   \n",
       "392    NC_002937   2305774      Δ2 bp          None             ?   \n",
       "397    NC_002937   2497231        +TC          None          None   \n",
       "398    NC_002937   2497231        +TC          None          None   \n",
       "404    NC_002937   3140594        +GA          None             ?   \n",
       "405    NC_002937   3140594        +GA          None             ?   \n",
       "545    NC_002937   2081507      Δ1 bp          None          None   \n",
       "\n",
       "    html_freq_300 html_freq_500 html_freq_780 html_freq_1000 ref_genome_y  \\\n",
       "5            None          100%          None           None    NC_002937   \n",
       "17           None          None          None           100%    NC_002937   \n",
       "28           None          None          None           100%    NC_002937   \n",
       "29           None          None          None           100%    NC_002937   \n",
       "32           None          None          None           100%    NC_002937   \n",
       "33           None          None          None           100%    NC_002937   \n",
       "46           None          100%          None           None    NC_005791   \n",
       "162          100%          100%          100%           100%    NC_002937   \n",
       "179          100%          None          100%           100%    NC_002937   \n",
       "190          None          None          None           100%    NC_002937   \n",
       "191          None          None          None           100%    NC_002937   \n",
       "196          None          None          None           100%    NC_002937   \n",
       "197          None          None          None           100%    NC_002937   \n",
       "271          100%          100%          None           100%    NC_002937   \n",
       "291          100%          None          100%           100%    NC_002937   \n",
       "300          None          None          None           100%    NC_002937   \n",
       "301          None          None          None           100%    NC_002937   \n",
       "302          None          None          None           100%    NC_002937   \n",
       "303          None          None          None           100%    NC_002937   \n",
       "325          None          100%          None           None    NC_005791   \n",
       "337          None          100%          None           None    NC_005791   \n",
       "354          None          None          None           None    NC_005863   \n",
       "386          None          None          100%           None    NC_002937   \n",
       "392          None          None          100%           100%    NC_002937   \n",
       "397          None          None          100%           100%    NC_002937   \n",
       "398          None          None          100%           100%    NC_002937   \n",
       "404          None          None          None           100%    NC_002937   \n",
       "405          None          None          None           100%    NC_002937   \n",
       "545          None          100%          None            NaN    NC_002937   \n",
       "\n",
       "    mutation_y line  generation  frequency  polymorphism_frequency new_cov  \\\n",
       "5          C→T  UA3       100.0   1.000000                0.818228     4/5   \n",
       "17         C→T  UA3       300.0   0.046179                0.046179     9/3   \n",
       "28         +GA  UA3       500.0   0.800007                0.800007   20/24   \n",
       "29         +GA  UA3       500.0   0.800014                0.800014   20/24   \n",
       "32         +GA  UA3       100.0   0.600002                0.600002     2/1   \n",
       "33         +GA  UA3       100.0   0.600006                0.600006     2/1   \n",
       "46         A→G  UA3       100.0   0.894805                0.894805   28/23   \n",
       "162         +G  UR1       100.0   0.926836                0.926836   17/21   \n",
       "179         +C  UR1       500.0   1.000000                0.934467     6/8   \n",
       "190        +CC  UR1       100.0   0.588240                0.588240     2/8   \n",
       "191        +CC  UR1       100.0   0.588240                0.588240     2/8   \n",
       "196        +CC  UR1       500.0   0.500004                0.500004     1/2   \n",
       "197        +CC  UR1       500.0   0.500004                0.500004     1/2   \n",
       "271         +C  US1       780.0   1.000000                0.923082     5/7   \n",
       "291      Δ2 bp  US1       500.0   1.000000                0.937500     6/9   \n",
       "300        +GA  US1       500.0   0.823488                0.823488     5/8   \n",
       "301        +GA  US1       500.0   0.764713                0.764713     5/8   \n",
       "302        +GA  US1       780.0   0.800004                0.800004   10/10   \n",
       "303        +GA  US1       780.0   0.800004                0.800004   10/10   \n",
       "325        G→A  US1       100.0   1.000000                0.882416    4/11   \n",
       "337        G→A  US1       100.0   0.894879                0.894879   17/17   \n",
       "354         +A  US1       500.0   1.000000                0.857157     3/3   \n",
       "386        C→T  HA3      1000.0   1.000000                0.942985   16/17   \n",
       "392      Δ2 bp  HA3       300.0   0.069762                0.069762     5/4   \n",
       "397        +TC  HA3       300.0   0.039215                0.039215     3/3   \n",
       "398        +TC  HA3       300.0   0.039215                0.039215     3/3   \n",
       "404        +GA  HA3       500.0   0.687505                0.687505   11/11   \n",
       "405        +GA  HA3       500.0   0.687508                0.687508   11/11   \n",
       "545      Δ1 bp  HR2       100.0   0.033324                0.033324     4/1   \n",
       "\n",
       "    total_cov     reject_reason evidence_type  item_id  \n",
       "5         4/7  FREQUENCY_CUTOFF            RA    111.0  \n",
       "17    126/133  FREQUENCY_CUTOFF            RA    481.0  \n",
       "28      29/26   STRAND_COVERAGE            RA    267.0  \n",
       "29      29/26   STRAND_COVERAGE            RA    268.0  \n",
       "32        2/3   STRAND_COVERAGE            RA    136.0  \n",
       "33        2/3   STRAND_COVERAGE            RA    137.0  \n",
       "46      31/26   STRAND_COVERAGE            RA    279.0  \n",
       "162     18/23   STRAND_COVERAGE            RA    227.0  \n",
       "179       6/9  FREQUENCY_CUTOFF            RA    174.0  \n",
       "190      6/11   STRAND_COVERAGE            RA    344.0  \n",
       "191      6/11   STRAND_COVERAGE            RA    345.0  \n",
       "196       3/3   STRAND_COVERAGE            RA    186.0  \n",
       "197       3/3   STRAND_COVERAGE            RA    187.0  \n",
       "271       5/8  FREQUENCY_CUTOFF            RA    212.0  \n",
       "291      6/10  FREQUENCY_CUTOFF            RA    167.0  \n",
       "300       8/9   STRAND_COVERAGE            RA    169.0  \n",
       "301       8/9   STRAND_COVERAGE            RA    170.0  \n",
       "302     13/12   STRAND_COVERAGE            RA    446.0  \n",
       "303     13/12   STRAND_COVERAGE            RA    447.0  \n",
       "325      5/12  FREQUENCY_CUTOFF            RA   1263.0  \n",
       "337     17/21   STRAND_COVERAGE            RA   1326.0  \n",
       "354       4/3  FREQUENCY_CUTOFF            RA   8236.0  \n",
       "386     17/18  FREQUENCY_CUTOFF            RA     88.0  \n",
       "392     78/51   STRAND_COVERAGE            RA    497.0  \n",
       "397     86/67  FREQUENCY_CUTOFF            RA    522.0  \n",
       "398     86/67  FREQUENCY_CUTOFF            RA    523.0  \n",
       "404     14/18   STRAND_COVERAGE            RA    386.0  \n",
       "405     14/18   STRAND_COVERAGE            RA    387.0  \n",
       "545     77/73  FREQUENCY_CUTOFF            RA    327.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#1b. Which ones?\n",
    "all_freqs_evidence.dropna(subset=['reject_reason'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_genome_x</th>\n",
       "      <th>position</th>\n",
       "      <th>mutation_x</th>\n",
       "      <th>html_freq_anc</th>\n",
       "      <th>html_freq_100</th>\n",
       "      <th>html_freq_300</th>\n",
       "      <th>html_freq_500</th>\n",
       "      <th>html_freq_780</th>\n",
       "      <th>html_freq_1000</th>\n",
       "      <th>ref_genome_y</th>\n",
       "      <th>mutation_y</th>\n",
       "      <th>line</th>\n",
       "      <th>generation</th>\n",
       "      <th>frequency</th>\n",
       "      <th>polymorphism_frequency</th>\n",
       "      <th>new_cov</th>\n",
       "      <th>total_cov</th>\n",
       "      <th>reject_reason</th>\n",
       "      <th>evidence_type</th>\n",
       "      <th>item_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>1376015</td>\n",
       "      <td>C→T</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>C→T</td>\n",
       "      <td>UA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.046179</td>\n",
       "      <td>0.046179</td>\n",
       "      <td>9/3</td>\n",
       "      <td>126/133</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>481.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2497231</td>\n",
       "      <td>+TC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+TC</td>\n",
       "      <td>HA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.039215</td>\n",
       "      <td>0.039215</td>\n",
       "      <td>3/3</td>\n",
       "      <td>86/67</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>522.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2497231</td>\n",
       "      <td>+TC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>+TC</td>\n",
       "      <td>HA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.039215</td>\n",
       "      <td>0.039215</td>\n",
       "      <td>3/3</td>\n",
       "      <td>86/67</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>523.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2081507</td>\n",
       "      <td>Δ1 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>Δ1 bp</td>\n",
       "      <td>HR2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.033324</td>\n",
       "      <td>0.033324</td>\n",
       "      <td>4/1</td>\n",
       "      <td>77/73</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "      <td>327.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ref_genome_x  position mutation_x html_freq_anc html_freq_100  \\\n",
       "17     NC_002937   1376015        C→T          None          None   \n",
       "397    NC_002937   2497231        +TC          None          None   \n",
       "398    NC_002937   2497231        +TC          None          None   \n",
       "545    NC_002937   2081507      Δ1 bp          None          None   \n",
       "\n",
       "    html_freq_300 html_freq_500 html_freq_780 html_freq_1000 ref_genome_y  \\\n",
       "17           None          None          None           100%    NC_002937   \n",
       "397          None          None          100%           100%    NC_002937   \n",
       "398          None          None          100%           100%    NC_002937   \n",
       "545          None          100%          None            NaN    NC_002937   \n",
       "\n",
       "    mutation_y line  generation  frequency  polymorphism_frequency new_cov  \\\n",
       "17         C→T  UA3       300.0   0.046179                0.046179     9/3   \n",
       "397        +TC  HA3       300.0   0.039215                0.039215     3/3   \n",
       "398        +TC  HA3       300.0   0.039215                0.039215     3/3   \n",
       "545      Δ1 bp  HR2       100.0   0.033324                0.033324     4/1   \n",
       "\n",
       "    total_cov     reject_reason evidence_type  item_id  \n",
       "17    126/133  FREQUENCY_CUTOFF            RA    481.0  \n",
       "397     86/67  FREQUENCY_CUTOFF            RA    522.0  \n",
       "398     86/67  FREQUENCY_CUTOFF            RA    523.0  \n",
       "545     77/73  FREQUENCY_CUTOFF            RA    327.0  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#2. switch to consensus mode because frequencies <5%, with known evidence?\n",
    "all_freqs_evidence[(all_freqs_evidence['polymorphism_frequency'] < 0.05) & (all_freqs_evidence['polymorphism_frequency'] > 0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(135, 20)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#How many '100%' frequencies are due to:\n",
    "#1. switch to consensus mode because frequencies >95%, with known evidence?\n",
    "polymorphism_over_95 = all_freqs_evidence[(all_freqs_evidence['polymorphism_frequency'] < 1) & (all_freqs_evidence['polymorphism_frequency'] > 0.95)]\n",
    "polymorphism_over_95.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "polymorphism_over_95"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(369, 20)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#2. actual read frequencies of 100%?\n",
    "polymorphism_100 = all_freqs_evidence[(all_freqs_evidence['polymorphism_frequency'] == 1) & ((all_freqs_evidence['reject_reason'] != 'FREQUENCY_CUTOFF') & (all_freqs_evidence['reject_reason'] != 'STRAND_COVERAGE'))]\n",
    "polymorphism_100.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "polymorphism_100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ref_genome_x              49\n",
       "position                  49\n",
       "mutation_x                49\n",
       "html_freq_anc             49\n",
       "html_freq_100             49\n",
       "html_freq_300             49\n",
       "html_freq_500             49\n",
       "html_freq_780             49\n",
       "html_freq_1000            31\n",
       "ref_genome_y              49\n",
       "mutation_y                49\n",
       "line                      49\n",
       "generation                49\n",
       "frequency                 49\n",
       "polymorphism_frequency    49\n",
       "new_cov                    0\n",
       "total_cov                  0\n",
       "reject_reason              0\n",
       "evidence_type             49\n",
       "item_id                   49\n",
       "dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#How many frequencies are due to unknown base evidence?\n",
    "all_freqs_evidence[all_freqs_evidence['evidence_type'] == 'UN'].count()\n",
    "#Not all evolution lines have been sequenced at 1,000 generations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2 Problems and Solutions\n",
    "\n",
    "#### Use of ('ref_genome', 'position') keys to match suspect frequencies with evidence type, misses extra genome positions on JC evidence columns\n",
    "\n",
    "JC (new junction) evidence for polymorphism (NC_002937, 629936) in line UR1 was not extracted from original annotated.gd file, because that specific (ref_genome, position) were in columns 6 (side_2_seq_id) and 7 (side_2_position). The (ref_genome, position) fields in the dataframe were extracted from columns 3 and 4, which are accurate for other evidence types, and for JC side_1_seq_id, and side_1_position.\n",
    "\n",
    "~~- Will eventually add code to extract side_2 information for JC evidence. In the meantime, UR1 (NC_002937, 629936) is noted, but not included in analyses until code is updated.~~\n",
    "- Solved, code added.\n",
    "\n",
    "#### Duplicate ('ref_genome', 'position') keys in frequency/evidence dictionaries are deleted (Python does not allow duplicate keys).\n",
    "\n",
    "- Solved by adding unique item_id to keys.\n",
    "\n",
    "#### 'frequency' field from annotated.gd is frequency reported after conversion to consensus mode.\n",
    "\n",
    "- Solved by adding polymorphism_frequency column, which reports actual frequencies in polymorphism mode."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
