{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identify procedural reasons for 'all-nothing-all' mutations (Part 1)\n",
    "## April 4, 2017\n",
    "\n",
    "Mutations that have 100% frequency in one generation, disappear completely (0%) in the next generation sampled, then reappear at 100% frequency. Potential explanations: \n",
    "\n",
    "1. Biological - Frequency-dependent selection? \n",
    "2. Procedural - Undersampled populations? Missing coverage during sequencing? Breseq frequency threshold for polymorphism mode (5-95%) not met?\n",
    "\n",
    "Input: \n",
    "1. HTML file from gdtools COMPARE\n",
    "2. GD files (annotated.gd) from breseq output.\n",
    "\n",
    "Tasks:\n",
    "1. Identify rows with 'all-nothing-all' (a-n-a) pattern in input HTML file, by mutation position.\n",
    "2. For each a-n-a, identify mutation position and generation(s) when 0% frequency occurs.\n",
    "3. For each 0% frequency occurrence in a-n-a, refer to relevant annotated.gd file to identify reason for frequency.\n",
    "4. Use prototype scripts here to build methods. See [lost_polymorphisms.py](https://github.com/yeemey/dvh_mms2/blob/master/lost_polymorphisms.py).\n",
    "\n",
    "#### Note:\n",
    "\n",
    "The following Part 1 prototype scripts have since been improved upon in [lost_polymorphisms.py](https://github.com/yeemey/dvh_mms2/blob/master/lost_polymorphisms.py) and no longer reflect what was used in analyses.\n",
    "\n",
    "```\n",
    "from bs4 import BeautifulSoup\n",
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "ua3 = BeautifulSoup(open('/Users/ymseah/Documents/compare/UA3.html'), 'html.parser')\n",
    "\n",
    "#Task 1-2\n",
    "\n",
    "#1. Identify line-generation headers\n",
    "line_generations = []\n",
    "next_line_gen = ua3.find('th', string=re.compile('0'))\n",
    "line_gen = next_line_gen.string\n",
    "while line_gen != 'annotation':\n",
    "    line_generations.append(line_gen)\n",
    "    next_line_gen = next_line_gen.next_element.next_element.next_element\n",
    "    line_gen = next_line_gen.string\n",
    "print(line_generations)\n",
    "\n",
    "#2. Identify ref genome and polymorphism position\n",
    "#3. Identify frequencies across generations\n",
    "#4. Link 2. and 3.\n",
    "\n",
    "table_cells = ua3.body.find_all('td')\n",
    "generation_frequencies_dict = {}\n",
    "for cell in table_cells:\n",
    "    if re.match('NC_', str(cell.string)):\n",
    "        position = cell.next_sibling.next_sibling.next_sibling\n",
    "        clean_position = int(re.sub(',', '', position.string))\n",
    "        gen_freqs_key = (cell.string, clean_position)\n",
    "        mutation = position.next_sibling.next_sibling.next_sibling\n",
    "        freq_ancestor = mutation.next_sibling.next_sibling.next_sibling\n",
    "        freq_gen100 = freq_ancestor.next_sibling\n",
    "        freq_gen300 = freq_gen100.next_sibling\n",
    "        freq_gen500 = freq_gen300.next_sibling\n",
    "        freq_gen780 = freq_gen500.next_sibling\n",
    "        freq_gen1000 = freq_gen780.next_sibling\n",
    "        gen_freqs_value = [mutation.string, freq_ancestor.string, freq_gen100.string, freq_gen300.string, freq_gen500.string, freq_gen780.string, freq_gen1000.string]\n",
    "        generation_frequencies_dict[gen_freqs_key] = gen_freqs_value\n",
    "\n",
    "#5. Identify a-n-a pattern.\n",
    "\n",
    "suspect_frequencies_dict = {}\n",
    "with open('suspect_frequencies.tsv', 'w') as output_file:\n",
    "    output_file.write('ref_genome\\tposition\\tmutation\\tfreq_anc\\tfreq_100\\tfreq_300\\tfreq_500\\tfreq_780\\tfreq_1000\\n')\n",
    "    for key, value in generation_frequencies_dict.items():\n",
    "        counter = 2\n",
    "        while counter <= len(value):\n",
    "            if counter + 1 < len(value):\n",
    "                if value[counter] == None:\n",
    "                    if value[counter - 1] == '100%' or value[counter + 1] == '100%':\n",
    "                        suspect_frequencies_dict[key] = value\n",
    "                        output_file.write(key[0] + '\\t' + str(key[1]) + '\\t')\n",
    "                        for item in value:\n",
    "                            output_file.write(str(item) + '\\t')\n",
    "                        output_file.write('\\n')\n",
    "                elif value[counter] == '100%':\n",
    "                    if value[counter - 1] == None or value[counter + 1] == None:\n",
    "                        suspect_frequencies_dict[key] = value\n",
    "                        output_file.write(key[0] + '\\t' + str(key[1]) + '\\t')\n",
    "                        for item in value:\n",
    "                            output_file.write(str(item) + '\\t')\n",
    "                        output_file.write('\\n')\n",
    "            counter += 2\n",
    "\n",
    "UA3_100 = pd.read_table('/Users/ymseah/Documents/sic_UA3-15_breseq/annotated.gd', comment='#', names=range(50), dtype=str)\n",
    "UA3_100.insert(0, 'generation', 100)\n",
    "UA3_300 = pd.read_table('/Users/ymseah/Documents/sic_UA3.45_breseq/annotated.gd', comment='#', names=range(50), dtype=str)\n",
    "UA3_300.insert(0, 'generation', 300)\n",
    "UA3_500 = pd.read_table('/Users/ymseah/Documents/sic_UA3-76_breseq/annotated.gd', comment='#', names=range(50), dtype=str)\n",
    "UA3_500.insert(0, 'generation', 500)\n",
    "UA3_780 = pd.read_table('/Users/ymseah/Documents/sic_UA3.118_breseq/annotated.gd', comment='#', names=range(50), dtype=str)\n",
    "UA3_780.insert(0, 'generation', 780)\n",
    "UA3_1000 = pd.read_table('/Users/ymseah/Documents/sic_UA3_S2_L001_breseq/output/evidence/annotated.gd', comment='#', names=range(50), dtype=str)\n",
    "UA3_1000.insert(0, 'generation', 1000)\n",
    "\n",
    "UA3_df  = pd.concat([UA3_100, UA3_300, UA3_500, UA3_780, UA3_1000], ignore_index=True)\n",
    "UA3_df.insert(0, 'line', 'UA3')\n",
    "UA3_df.insert(2, 'frequency', 0.0)\n",
    "UA3_df.insert(3, 'gene_product', '')\n",
    "UA3_df.insert(4, 'gene_position', '')\n",
    "UA3_df.insert(5, 'reject', '')\n",
    "\n",
    "for row in UA3_df.itertuples():\n",
    "    #check each column\n",
    "    col_index = 6\n",
    "    while col_index < 50:\n",
    "        #1. polymorphism frequencies\n",
    "        if re.match('frequency=', str(UA3_df.loc[row[0], col_index])):\n",
    "            UA3_df.loc[row[0], 'frequency'] = re.sub('frequency=', '', str(UA3_df.loc[row[0], col_index]))\n",
    "        #2. gene products\n",
    "        elif re.match('gene_product=', str(UA3_df.loc[row[0], col_index])):\n",
    "            UA3_df.loc[row[0], 'gene_product'] = re.sub('gene_product=', '', str(UA3_df.loc[row[0], col_index]))\n",
    "        #3. polymorphism rejection reasons\n",
    "        elif re.match('reject=', str(UA3_df.loc[row[0], col_index])):\n",
    "            UA3_df.loc[row[0], 'reject'] = re.sub('reject=', '', str(UA3_df.loc[row[0], col_index]))\n",
    "        #4. gene annotations\n",
    "        elif re.match('gene_position=', str(UA3_df.loc[row[0], col_index])):\n",
    "            UA3_df.loc[row[0], 'gene_position'] = re.sub('gene_position=', '', str(UA3_df.loc[row[0], col_index]))\n",
    "        col_index += 1\n",
    "    #set frequencies type to float\n",
    "    if re.match('1|2|3|4|5|6|7|8|9', str(UA3_df.loc[row[0], 'frequency'])):\n",
    "        UA3_df.loc[row[0], 'frequency'] = float(UA3_df.loc[row[0], 'frequency'])\n",
    "    else:\n",
    "        UA3_df.loc[row[0], 'frequency'] = 0.0\n",
    "    #set positions (col 4) type to int\n",
    "    UA3_df.loc[row[0], 4] = int(UA3_df.loc[row[0], 4])\n",
    "    #set reject col to 'NA' when no reject reason given.\n",
    "    if (UA3_df.loc[row[0], 'reject'] == '') & (UA3_df.loc[row[0], 2] == '.'):\n",
    "        UA3_df.loc[row[0], 'reject'] = 'NA'\n",
    "\n",
    "UA3_df.rename(columns = {0: 'entry_type', 1: 'item_id', 2: 'evidence_ids', 3: 'ref_genome', 4:'position'}, inplace=True)\n",
    "ua3df_subset = UA3_df[['line', 'generation', 'frequency', 'gene_product', 'gene_position', 'reject', 'entry_type', 'item_id', 'evidence_ids', 'ref_genome', 'position']].copy()\n",
    "ua3df_subset.to_csv('/Users/ymseah/Documents/ua3.csv', index=False)\n",
    "\n",
    "# Task 3.\n",
    "#6: Identify reasons for a-n-a frequencies.\n",
    "\n",
    "ua3df_subset_mutations_only = ua3df_subset[(ua3df_subset['entry_type'] == 'SNP') | \n",
    "                                           (ua3df_subset['entry_type'] == 'SUB') | \n",
    "                                           (ua3df_subset['entry_type'] == 'DEL') | \n",
    "                                           (ua3df_subset['entry_type'] == 'INS') | \n",
    "                                           (ua3df_subset['entry_type'] == 'MOB') | \n",
    "                                           (ua3df_subset['entry_type'] == 'AMP') | \n",
    "                                           (ua3df_subset['entry_type'] == 'CON') | \n",
    "                                           (ua3df_subset['entry_type'] == 'INV')]\n",
    "for key, value in suspect_frequencies_dict.items():\n",
    "    row_indices = ua3df_subset_mutations_only[(ua3df_subset_mutations_only['ref_genome'] == key[0]) & \n",
    "                                              (ua3df_subset_mutations_only['position'] == key[1])].index.tolist()\n",
    "    #print(key, row_indices)\n",
    "    for row in row_indices:\n",
    "        reject_reason = ua3df_subset_mutations_only.loc[row, 'reject']\n",
    "        generation = ua3df_subset_mutations_only.loc[row, 'generation']\n",
    "        entry_type = ua3df_subset_mutations_only.loc[row, 'entry_type']\n",
    "        #print(row, reject_reason, generation, entry_type)\n",
    "        if reject_reason != '':\n",
    "            if generation == 100:\n",
    "                value[2] = str(value[2]) + ' ' + reject_reason\n",
    "            elif generation == 300:\n",
    "                value[3] = str(value[3]) + ' ' + reject_reason\n",
    "            elif generation == 500:\n",
    "                value[4] = str(value[4]) + ' ' + reject_reason\n",
    "            elif generation == 780:\n",
    "                value[5] = str(value[5]) + ' ' + reject_reason\n",
    "            elif generation == 1000:\n",
    "                value[6] = str(value[6]) + ' ' + reject_reason\n",
    "```\n",
    "\n",
    "## Part 1 Problems & Solutions\n",
    "\n",
    "#### String sorting messes up generation order. \n",
    "\n",
    "```\n",
    "find_line_generations = set(ua3.find_all('th', string=re.compile('UA3')))\n",
    "line_generations=['0']\n",
    "for each in find_line_generations:\n",
    "    line_generations.append(each.string)\n",
    "    line_generations.sort()\n",
    "print(line_generations)\n",
    "```\n",
    "\n",
    "- Solved by appending in order of appearance in document.\n",
    "\n",
    "\n",
    "#### Identifying frequencies in each row, when there are no string elements within each cell to represent 0% frequency. \n",
    "\n",
    "\n",
    "```\n",
    "frequencies_by_generations = {}\n",
    "all_comments = ua3.find_all(text = lambda text:isinstance(text, Comment))\n",
    "for comment in all_comments:\n",
    "    if comment == ' Seq_Id ':\n",
    "        ref_genome = comment.previous_element.strip()\n",
    "        position = comment.next_element.next_element.next_element.strip()\n",
    "        genome_position_key = (ref_genome, position)\n",
    "        generation_frequencies_value = []\n",
    "        #gen_freq = comment.next_sibling.next_sibling.next_sibling.next_sibling.next_sibling.next_sibling.next_sibling.next_sibling.next_element.strip()\n",
    "        print(genome_position_key)\n",
    "```\n",
    "\n",
    "- Solved by listing each cell (td) as an element and searching within td. Initially, searching was by html comment position, but it got messy.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identify procedural reasons for 'all-nothing-all' mutations (Part 2)\n",
    "## April 24, 2017\n",
    "\n",
    "Tasks:\n",
    "1. Use output from [lost_polymorphisms_runScript.py](https://github.com/yeemey/dvh_mms2/blob/master/lost_polymorphisms_runScript.py) to classify procedural reasons, summary counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "ua3_freqs = pd.read_csv('~/Documents/compare/parse_compare/UA3_suspect_frequencies.tsv', sep='\\t')\n",
    "ua3_evidence = pd.read_csv('~/Documents/compare/parse_compare/UA3_rejected_evidence.tsv', sep='\\t')\n",
    "ua3freqs_evidence = pd.merge(ua3_freqs, ua3_evidence, how='left', on='position')\n",
    "\n",
    "ue3_freqs = pd.read_csv('~/Documents/compare/parse_compare/UE3_suspect_frequencies.tsv', sep='\\t')\n",
    "ue3_evidence = pd.read_csv('~/Documents/compare/parse_compare/UE3_rejected_evidence.tsv', sep='\\t')\n",
    "ue3freqs_evidence = pd.merge(ue3_freqs, ue3_evidence, how='left', on='position')\n",
    "\n",
    "ur1_freqs = pd.read_csv('~/Documents/compare/parse_compare/UR1_suspect_frequencies.tsv', sep='\\t')\n",
    "ur1_evidence = pd.read_csv('~/Documents/compare/parse_compare/UR1_rejected_evidence.tsv', sep='\\t')\n",
    "ur1freqs_evidence = pd.merge(ur1_freqs, ur1_evidence, how='left', on='position')\n",
    "\n",
    "us1_freqs = pd.read_csv('~/Documents/compare/parse_compare/US1_suspect_frequencies.tsv', sep='\\t')\n",
    "us1_evidence = pd.read_csv('~/Documents/compare/parse_compare/US1_rejected_evidence.tsv', sep='\\t')\n",
    "us1freqs_evidence = pd.merge(us1_freqs, us1_evidence, how='left', on='position')\n",
    "\n",
    "ha3_freqs = pd.read_csv('~/Documents/compare/parse_compare/HA3_suspect_frequencies.tsv', sep='\\t')\n",
    "ha3_evidence = pd.read_csv('~/Documents/compare/parse_compare/HA3_rejected_evidence.tsv', sep='\\t')\n",
    "ha3freqs_evidence = pd.merge(ha3_freqs, ha3_evidence, how='left', on='position')\n",
    "\n",
    "he3_freqs = pd.read_csv('~/Documents/compare/parse_compare/HE3_suspect_frequencies.tsv', sep='\\t')\n",
    "he3_evidence = pd.read_csv('~/Documents/compare/parse_compare/HE3_rejected_evidence.tsv', sep='\\t')\n",
    "he3freqs_evidence = pd.merge(he3_freqs, he3_evidence, how='left', on='position')\n",
    "\n",
    "hr2_freqs = pd.read_csv('~/Documents/compare/parse_compare/HR2_suspect_frequencies.tsv', sep='\\t')\n",
    "hr2_evidence = pd.read_csv('~/Documents/compare/parse_compare/HR2_rejected_evidence.tsv', sep='\\t')\n",
    "hr2freqs_evidence = pd.merge(hr2_freqs, hr2_evidence, how='left', on='position')\n",
    "\n",
    "hs3_freqs = pd.read_csv('~/Documents/compare/parse_compare/HS3_suspect_frequencies.tsv', sep='\\t')\n",
    "hs3_evidence = pd.read_csv('~/Documents/compare/parse_compare/HS3_rejected_evidence.tsv', sep='\\t')\n",
    "hs3freqs_evidence = pd.merge(hs3_freqs, hs3_evidence, how='left', on='position')\n",
    "\n",
    "#All results with suspect frequencies, i.e., None --> 100%, or 100% --> None\n",
    "#From COMPARE html output: ref_genome_x, position, mutation, html_freq__\n",
    "#From annotated.gd: ref_genome_y, line, generation, frequency, reject_reason, evidence_type\n",
    "all_freqs_evidence = pd.concat([ua3freqs_evidence, ue3freqs_evidence, ur1freqs_evidence, us1freqs_evidence, ha3freqs_evidence, he3freqs_evidence, hr2freqs_evidence, hs3freqs_evidence], ignore_index=True)\n",
    "all_freqs_evidence.to_csv('~/Documents/compare/parse_compare/all_freqs_evidence.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_genome_x</th>\n",
       "      <th>position</th>\n",
       "      <th>mutation</th>\n",
       "      <th>html_freq_anc</th>\n",
       "      <th>html_freq_100</th>\n",
       "      <th>html_freq_300</th>\n",
       "      <th>html_freq_500</th>\n",
       "      <th>html_freq_780</th>\n",
       "      <th>html_freq_1000</th>\n",
       "      <th>ref_genome_y</th>\n",
       "      <th>line</th>\n",
       "      <th>generation</th>\n",
       "      <th>frequency</th>\n",
       "      <th>evidence_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reject_reason</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FREQUENCY_CUTOFF</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRAND_COVERAGE</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ref_genome_x  position  mutation  html_freq_anc  \\\n",
       "reject_reason                                                       \n",
       "FREQUENCY_CUTOFF             4         4         4              4   \n",
       "STRAND_COVERAGE              2         2         2              2   \n",
       "\n",
       "                  html_freq_100  html_freq_300  html_freq_500  html_freq_780  \\\n",
       "reject_reason                                                                  \n",
       "FREQUENCY_CUTOFF              4              4              4              4   \n",
       "STRAND_COVERAGE               2              2              2              2   \n",
       "\n",
       "                  html_freq_1000  ref_genome_y  line  generation  frequency  \\\n",
       "reject_reason                                                                 \n",
       "FREQUENCY_CUTOFF               4             4     4           4          4   \n",
       "STRAND_COVERAGE                2             2     2           2          2   \n",
       "\n",
       "                  evidence_type  \n",
       "reject_reason                    \n",
       "FREQUENCY_CUTOFF              4  \n",
       "STRAND_COVERAGE               2  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#How many 'None' frequencies are due to:\n",
    "#1a. rejection based on evidence filters?\n",
    "all_freqs_evidence.groupby('reject_reason').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_genome_x</th>\n",
       "      <th>position</th>\n",
       "      <th>mutation</th>\n",
       "      <th>html_freq_anc</th>\n",
       "      <th>html_freq_100</th>\n",
       "      <th>html_freq_300</th>\n",
       "      <th>html_freq_500</th>\n",
       "      <th>html_freq_780</th>\n",
       "      <th>html_freq_1000</th>\n",
       "      <th>ref_genome_y</th>\n",
       "      <th>line</th>\n",
       "      <th>generation</th>\n",
       "      <th>frequency</th>\n",
       "      <th>reject_reason</th>\n",
       "      <th>evidence_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>708564</td>\n",
       "      <td>C→T</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>UA3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_005791</td>\n",
       "      <td>540893</td>\n",
       "      <td>A→G</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005791</td>\n",
       "      <td>UA3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.894805</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2083305</td>\n",
       "      <td>+C</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>UR1</td>\n",
       "      <td>500.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>30836</td>\n",
       "      <td>+A</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>US1</td>\n",
       "      <td>500.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2305774</td>\n",
       "      <td>Δ2 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>?</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>HA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.069762</td>\n",
       "      <td>STRAND_COVERAGE</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2497231</td>\n",
       "      <td>+TC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>HA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.039215</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ref_genome_x  position mutation html_freq_anc html_freq_100 html_freq_300  \\\n",
       "3      NC_002937    708564      C→T          None          None          None   \n",
       "13     NC_005791    540893      A→G          None          None          None   \n",
       "64     NC_002937   2083305       +C          100%          100%          100%   \n",
       "109    NC_005863     30836       +A          100%          100%          None   \n",
       "126    NC_002937   2305774    Δ2 bp          None             ?          None   \n",
       "127    NC_002937   2497231      +TC          None          None          None   \n",
       "\n",
       "    html_freq_500 html_freq_780 html_freq_1000 ref_genome_y line  generation  \\\n",
       "3            100%          None           None    NC_002937  UA3       100.0   \n",
       "13           100%          None           None    NC_005791  UA3       100.0   \n",
       "64           None          100%           100%    NC_002937  UR1       500.0   \n",
       "109          None          None           None    NC_005863  US1       500.0   \n",
       "126          None          100%           100%    NC_002937  HA3       300.0   \n",
       "127          None          100%           100%    NC_002937  HA3       300.0   \n",
       "\n",
       "     frequency     reject_reason evidence_type  \n",
       "3     1.000000  FREQUENCY_CUTOFF            RA  \n",
       "13    0.894805   STRAND_COVERAGE            RA  \n",
       "64    1.000000  FREQUENCY_CUTOFF            RA  \n",
       "109   1.000000  FREQUENCY_CUTOFF            RA  \n",
       "126   0.069762   STRAND_COVERAGE            RA  \n",
       "127   0.039215  FREQUENCY_CUTOFF            RA  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#1b. Which ones?\n",
    "all_freqs_evidence.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_genome_x</th>\n",
       "      <th>position</th>\n",
       "      <th>mutation</th>\n",
       "      <th>html_freq_anc</th>\n",
       "      <th>html_freq_100</th>\n",
       "      <th>html_freq_300</th>\n",
       "      <th>html_freq_500</th>\n",
       "      <th>html_freq_780</th>\n",
       "      <th>html_freq_1000</th>\n",
       "      <th>ref_genome_y</th>\n",
       "      <th>line</th>\n",
       "      <th>generation</th>\n",
       "      <th>frequency</th>\n",
       "      <th>reject_reason</th>\n",
       "      <th>evidence_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>99978</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>19.7%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>UA3</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NC_005791</td>\n",
       "      <td>332634</td>\n",
       "      <td>Δ4 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_005791</td>\n",
       "      <td>UE3</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>99978</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>UE3</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>16512</td>\n",
       "      <td>Δ3 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>66.0%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>UR1</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>NC_005791</td>\n",
       "      <td>1172630</td>\n",
       "      <td>Δ189 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005791</td>\n",
       "      <td>UR1</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>99978</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>UR1</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>1930199</td>\n",
       "      <td>Δ69 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>US1</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>1</td>\n",
       "      <td>Δ202,301 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>US1</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>99978</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>US1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>1304113</td>\n",
       "      <td>Δ15 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>?</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>HA3</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2497231</td>\n",
       "      <td>+TC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>HA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.039215</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3535553</td>\n",
       "      <td>Δ28 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>?</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>HA3</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>99978</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>HA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>NC_005791</td>\n",
       "      <td>332668</td>\n",
       "      <td>Δ3 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>24.1%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_005791</td>\n",
       "      <td>HE3</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>1</td>\n",
       "      <td>Δ202,301 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>HE3</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>99978</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>?</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>HE3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>885166</td>\n",
       "      <td>Δ3 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>HR2</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>1187321</td>\n",
       "      <td>Δ12 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>HR2</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>3396071</td>\n",
       "      <td>Δ28 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>HR2</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>1</td>\n",
       "      <td>Δ202,301 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>HR2</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>99978</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>HR2</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>885182</td>\n",
       "      <td>Δ3 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>?</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>HS3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>1</td>\n",
       "      <td>Δ202,301 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>HS3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>99978</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>HS3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ref_genome_x  position     mutation html_freq_anc html_freq_100  \\\n",
       "28     NC_005863     99978       Δ21 bp          100%          None   \n",
       "39     NC_005791    332634        Δ4 bp          None          None   \n",
       "54     NC_005863     99978       Δ21 bp          100%          100%   \n",
       "56     NC_002937     16512        Δ3 bp          None          None   \n",
       "75     NC_005791   1172630      Δ189 bp          None          None   \n",
       "84     NC_005863     99978       Δ21 bp          100%          None   \n",
       "90     NC_002937   1930199       Δ69 bp          None          None   \n",
       "108    NC_005863         1  Δ202,301 bp          None          None   \n",
       "111    NC_005863     99978       Δ21 bp          100%          100%   \n",
       "121    NC_002937   1304113       Δ15 bp          None             ?   \n",
       "127    NC_002937   2497231          +TC          None          None   \n",
       "129    NC_002937   3535553       Δ28 bp          None             ?   \n",
       "145    NC_005863     99978       Δ21 bp          100%          None   \n",
       "154    NC_005791    332668        Δ3 bp          None          None   \n",
       "161    NC_005863         1  Δ202,301 bp          None          None   \n",
       "164    NC_005863     99978       Δ21 bp          100%          100%   \n",
       "169    NC_002937    885166        Δ3 bp          None          None   \n",
       "170    NC_002937   1187321       Δ12 bp          None          None   \n",
       "175    NC_002937   3396071       Δ28 bp          None          None   \n",
       "190    NC_005863         1  Δ202,301 bp          None          None   \n",
       "193    NC_005863     99978       Δ21 bp          100%          None   \n",
       "198    NC_002937    885182        Δ3 bp          None          None   \n",
       "220    NC_005863         1  Δ202,301 bp          None          None   \n",
       "223    NC_005863     99978       Δ21 bp          100%          100%   \n",
       "\n",
       "    html_freq_300 html_freq_500 html_freq_780 html_freq_1000 ref_genome_y  \\\n",
       "28          19.7%          None          100%           100%    NC_005863   \n",
       "39           None          None          100%            NaN    NC_005791   \n",
       "54           100%          None          100%            NaN    NC_005863   \n",
       "56          66.0%          None          100%           100%    NC_002937   \n",
       "75           None          100%          None           None    NC_005791   \n",
       "84           100%          None          100%           100%    NC_005863   \n",
       "90           100%          None          100%           100%    NC_002937   \n",
       "108          100%          None          100%           None    NC_005863   \n",
       "111          None          None          None           None    NC_005863   \n",
       "121          100%          None          100%           100%    NC_002937   \n",
       "127          None          None          100%           100%    NC_002937   \n",
       "129          None          100%          None           None    NC_002937   \n",
       "145          100%          None          100%           100%    NC_005863   \n",
       "154         24.1%          None          100%            NaN    NC_005791   \n",
       "161          100%          None          100%            NaN    NC_005863   \n",
       "164          None             ?          None            NaN    NC_005863   \n",
       "169          100%          100%          100%            NaN    NC_002937   \n",
       "170          100%          None          100%            NaN    NC_002937   \n",
       "175          100%          None          100%            NaN    NC_002937   \n",
       "190          None          100%          None            NaN    NC_005863   \n",
       "193          100%          None          100%            NaN    NC_005863   \n",
       "198          100%             ?          None           None    NC_002937   \n",
       "220          100%          None          None           None    NC_005863   \n",
       "223          None          None          100%           None    NC_005863   \n",
       "\n",
       "    line  generation  frequency     reject_reason evidence_type  \n",
       "28   UA3      1000.0   0.000000               NaN            UN  \n",
       "39   UE3       780.0   0.000000               NaN            UN  \n",
       "54   UE3       780.0   0.000000               NaN            UN  \n",
       "56   UR1       780.0   0.000000               NaN            UN  \n",
       "75   UR1       500.0   0.000000               NaN            UN  \n",
       "84   UR1       780.0   0.000000               NaN            UN  \n",
       "90   US1      1000.0   0.000000               NaN            UN  \n",
       "108  US1      1000.0   0.000000               NaN            UN  \n",
       "111  US1       100.0   0.000000               NaN            UN  \n",
       "121  HA3       780.0   0.000000               NaN            UN  \n",
       "127  HA3       300.0   0.039215  FREQUENCY_CUTOFF            RA  \n",
       "129  HA3       500.0   0.000000               NaN            UN  \n",
       "145  HA3       300.0   0.000000               NaN            UN  \n",
       "154  HE3       780.0   0.000000               NaN            UN  \n",
       "161  HE3       780.0   0.000000               NaN            MC  \n",
       "164  HE3       100.0   0.000000               NaN            UN  \n",
       "169  HR2       500.0   0.000000               NaN            UN  \n",
       "170  HR2       780.0   0.000000               NaN            UN  \n",
       "175  HR2       780.0   0.000000               NaN            UN  \n",
       "190  HR2       500.0   0.000000               NaN            MC  \n",
       "193  HR2       780.0   0.000000               NaN            UN  \n",
       "198  HS3       300.0   0.000000               NaN            UN  \n",
       "220  HS3       300.0   0.000000               NaN            MC  \n",
       "223  HS3       100.0   0.000000               NaN            UN  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#2. switch to consensus mode because frequencies <5%?\n",
    "all_freqs_evidence[all_freqs_evidence['frequency'] < 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_genome_x</th>\n",
       "      <th>position</th>\n",
       "      <th>mutation</th>\n",
       "      <th>html_freq_anc</th>\n",
       "      <th>html_freq_100</th>\n",
       "      <th>html_freq_300</th>\n",
       "      <th>html_freq_500</th>\n",
       "      <th>html_freq_780</th>\n",
       "      <th>html_freq_1000</th>\n",
       "      <th>ref_genome_y</th>\n",
       "      <th>line</th>\n",
       "      <th>generation</th>\n",
       "      <th>frequency</th>\n",
       "      <th>reject_reason</th>\n",
       "      <th>evidence_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>NC_002937</td>\n",
       "      <td>2497231</td>\n",
       "      <td>+TC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>100%</td>\n",
       "      <td>NC_002937</td>\n",
       "      <td>HA3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.039215</td>\n",
       "      <td>FREQUENCY_CUTOFF</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>1</td>\n",
       "      <td>Δ202,301 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>HE3</td>\n",
       "      <td>780.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>1</td>\n",
       "      <td>Δ202,301 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>HR2</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>NC_005863</td>\n",
       "      <td>1</td>\n",
       "      <td>Δ202,301 bp</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100%</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_005863</td>\n",
       "      <td>HS3</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ref_genome_x  position     mutation html_freq_anc html_freq_100  \\\n",
       "127    NC_002937   2497231          +TC          None          None   \n",
       "161    NC_005863         1  Δ202,301 bp          None          None   \n",
       "190    NC_005863         1  Δ202,301 bp          None          None   \n",
       "220    NC_005863         1  Δ202,301 bp          None          None   \n",
       "\n",
       "    html_freq_300 html_freq_500 html_freq_780 html_freq_1000 ref_genome_y  \\\n",
       "127          None          None          100%           100%    NC_002937   \n",
       "161          100%          None          100%            NaN    NC_005863   \n",
       "190          None          100%          None            NaN    NC_005863   \n",
       "220          100%          None          None           None    NC_005863   \n",
       "\n",
       "    line  generation  frequency     reject_reason evidence_type  \n",
       "127  HA3       300.0   0.039215  FREQUENCY_CUTOFF            RA  \n",
       "161  HE3       780.0   0.000000               NaN            MC  \n",
       "190  HR2       500.0   0.000000               NaN            MC  \n",
       "220  HS3       300.0   0.000000               NaN            MC  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#3. switch to consensus mode because frequencies <5%?, with known evidence?\n",
    "all_freqs_evidence[(all_freqs_evidence['frequency'] < 0.05) & (all_freqs_evidence['evidence_type'] != 'UN')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2 Problems and Solutions\n",
    "\n",
    "#### Use of 'ref_genome' and 'position' to match suspect frequencies with evidence type\n",
    "\n",
    "JC (new junction) evidence for polymorphism (NC_002937, 629936) in line UR1 was not extracted from original annotated.gd file, because that specific (ref_genome, position) were in columns 6 (side_2_seq_id) and 7 (side_2_position). The (ref_genome, position) fields in the dataframe were extracted from columns 3 and 4, which are accurate for other evidence types, and for JC side_1_seq_id, and side_1_position.\n",
    "\n",
    "~~- Will eventually add code to extract side_2 information for JC evidence. In the meantime, UR1 (NC_002937, 629936) is noted, but not included in analyses until code is updated.~~\n",
    "- Solved, code added.\n",
    "\n",
    "#### Duplicate keys in frequency/evidence dictionaries are deleted (Python does not allow duplicate keys)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
