{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sum_cov(cov_str):\n",
    "    cov_ints = cov_str.split('/')\n",
    "    total_cov = 0\n",
    "    for cov in cov_ints:\n",
    "        total_cov += int(cov)\n",
    "    return total_cov\n",
    "\n",
    "def subset_gd_to_df(gdfile, cov=False):\n",
    "    '''\n",
    "    Returns dataframe created from annotated.gd. All rows are preserved,\n",
    "    but only selected variables, namely entry type, entry id, \n",
    "    evidence id, genome id, position, mutation detail, frequency, gene product, \n",
    "    reject reasons, prediction mode, and polymorphism frequencies. \n",
    "    \n",
    "    If cov=True, will also report total coverage (i.e., total_cov), \n",
    "    RA coverage (new_cov), JC coverage (new_junction_read_count), \n",
    "    and MC-flanking coverage (left_outside_cov + right_outside_cov).\n",
    "    '''\n",
    "    df = pd.read_table(gdfile, comment='#', names=range(200), dtype=str)\n",
    "    df = df.dropna(axis=1, how='all')\n",
    "    num_columns = len(df.columns)\n",
    "    # https://stackoverflow.com/questions/27700591/reading-csv-files-with-messy-structure-with-pandas\n",
    "    df.rename(columns = {0: 'entry_type', 1: 'entry_id', 2: 'evidence_id', \n",
    "                         3: 'genome_id', 4: 'position', 5: 'mutation_detail'}, inplace=True)\n",
    "    for row in df.index:\n",
    "        #check each column\n",
    "        col_index = 6\n",
    "        while col_index < num_columns:\n",
    "            #1. mutation frequencies\n",
    "            if re.match('frequency=', str(df.loc[row, col_index])):\n",
    "                df.loc[row, 'frequency'] = re.sub('frequency=', '', str(df.loc[row, col_index]))\n",
    "                if df.loc[row, 'frequency'] == 'NA':\n",
    "                    df.loc[row, 'frequency'] = np.nan\n",
    "            #2. gene products\n",
    "            elif re.match('gene_product=', str(df.loc[row, col_index])):\n",
    "                df.loc[row, 'gene_product'] = re.sub('gene_product=', '', str(df.loc[row, col_index]))\n",
    "            #3. polymorphism rejection reasons\n",
    "            elif re.match('reject=', str(df.loc[row, col_index])):\n",
    "                df.loc[row, 'reject'] = re.sub('reject=', '', str(df.loc[row, col_index]))\n",
    "            #4. prediction type\n",
    "            elif re.match('prediction=', str(df.loc[row, col_index])):\n",
    "                df.loc[row, 'prediction'] = re.sub('prediction=', '', str(df.loc[row, col_index]))\n",
    "            #5. polymorphism mode frequencies\n",
    "            elif re.match('polymorphism_frequency=', str(df.loc[row, col_index])):\n",
    "                df.loc[row, 'polymorphism_frequency'] = re.sub('polymorphism_frequency=', '', str(df.loc[row, col_index]))\n",
    "                if df.loc[row, 'polymorphism_frequency'] == 'NA':\n",
    "                    df.loc[row, 'polymorphism_frequency'] = np.nan\n",
    "                    \n",
    "            if cov == True:\n",
    "                if re.match('major_cov=', str(df.loc[row, col_index])):\n",
    "                    major_cov = re.sub('major_cov=', '', str(df.loc[row, col_index]))\n",
    "                    df.loc[row, 'major_cov'] = sum_cov(major_cov)\n",
    "                elif re.match('minor_cov', str(df.loc[row, col_index])):\n",
    "                    minor_cov = re.sub('minor_cov=', '', str(df.loc[row, col_index]))\n",
    "                    df.loc[row, 'minor_cov'] = sum_cov(minor_cov)\n",
    "                elif re.match('total_cov=', str(df.loc[row, col_index])):\n",
    "                    total_cov = re.sub('total_cov=', '', str(df.loc[row, col_index]))\n",
    "                    df.loc[row, 'total_cov'] = sum_cov(total_cov)\n",
    "                elif re.match('new_cov=', str(df.loc[row, col_index])):\n",
    "                    ra_cov = re.sub('new_cov=', '', str(df.loc[row, col_index]))\n",
    "                    df.loc[row, 'ra_cov'] = sum_cov(ra_cov)\n",
    "                elif re.match('new_junction_read_count=', str(df.loc[row, col_index])):\n",
    "                    df.loc[row, 'jc_cov'] = re.sub('new_junction_read_count=', '', str(df.loc[row, col_index]))\n",
    "                elif re.match('left_outside_cov=', str(df.loc[row, col_index])):\n",
    "                    left_cov = re.sub('left_outside_cov=', '', str(df.loc[row, col_index]))\n",
    "                    if left_cov == 'NA':\n",
    "                        left_cov = 0\n",
    "                    else:\n",
    "                        df.loc[row, 'left_cov'] = int(left_cov)\n",
    "                elif re.match('right_outside_cov', str(df.loc[row, col_index])):\n",
    "                    right_cov = re.sub('right_outside_cov=', '', str(df.loc[row, col_index]))\n",
    "                    if right_cov == 'NA':\n",
    "                        right_cov = 0\n",
    "                    else:\n",
    "                        df.loc[row, 'right_cov'] = int(right_cov)\n",
    "            col_index += 1\n",
    "        #set reject col to 'NA' when no reject reason given.\n",
    "        if 'reject' in df.columns:\n",
    "            if (df.loc[row, 'reject'] == '') & (df.loc[row, 'evidence_id'] == '.'):\n",
    "                df.loc[row, 'reject'] = np.nan\n",
    "        else:\n",
    "            df['reject'] = np.nan\n",
    "    #set frequencies type to float \n",
    "    df[['frequency', 'polymorphism_frequency']] = df[['frequency', 'polymorphism_frequency']].astype(float)\n",
    "    if cov == True:\n",
    "        df[['left_cov', 'right_cov']].fillna(0)\n",
    "        df['mc_cov'] = df.left_cov + df.right_cov\n",
    "        return df[['entry_type', 'entry_id', 'evidence_id', 'genome_id', 'position', 'mutation_detail', \n",
    "                   'frequency', 'gene_product', 'reject', 'prediction', 'polymorphism_frequency', \n",
    "                   'major_cov', 'minor_cov', 'total_cov', 'ra_cov', 'jc_cov', 'mc_cov']].copy()\n",
    "    else:\n",
    "        return df[['entry_type', 'entry_id', 'evidence_id', 'genome_id', 'position', 'mutation_detail', \n",
    "                    'frequency', 'gene_product', 'reject', 'prediction', 'polymorphism_frequency']].copy()\n",
    "\n",
    "def select_mutation_rows(df):\n",
    "    '''\n",
    "    Selects mutation entry rows from dataframe. \n",
    "    '''\n",
    "    df_mutations = df[(df['entry_type'] == 'INS') | \n",
    "                      (df['entry_type'] == 'DEL') | \n",
    "                      (df['entry_type'] == 'SNP') | \n",
    "                      (df['entry_type'] == 'SUB') | \n",
    "                      (df['entry_type'] == 'MOB') | \n",
    "                      (df['entry_type'] == 'AMP') | \n",
    "                      (df['entry_type'] == 'CON') | \n",
    "                      (df['entry_type'] == 'INV')]\n",
    "    return df_mutations\n",
    "\n",
    "def add_evidence_to_mutation_rows(df, cov=False):\n",
    "    '''\n",
    "    Adds new column for evidence. \n",
    "    Evidence type taken from evidence entry rows, added to mutation entry rows.\n",
    "    '''\n",
    "    df_mutations = select_mutation_rows(df)\n",
    "    df['evidence'] = ''\n",
    "    for evidence in df_mutations['evidence_id']:\n",
    "        multi_evidence = evidence.split(',')\n",
    "        count = 0\n",
    "        while count < len(multi_evidence):\n",
    "            entry_row_index = df_mutations[df_mutations['evidence_id'] == evidence].index[0]\n",
    "            evidence_row_index = df[df['entry_id'] == multi_evidence[count]].index[0]\n",
    "            df.loc[entry_row_index, 'prediction'] = df.loc[evidence_row_index, 'prediction']\n",
    "            df.loc[entry_row_index, 'polymorphism_frequency'] = df.loc[evidence_row_index, 'polymorphism_frequency']\n",
    "            df.loc[entry_row_index, 'evidence'] = df.loc[entry_row_index, 'evidence'] + df.loc[evidence_row_index, 'entry_type']\n",
    "            if cov == True:\n",
    "                df.loc[entry_row_index, 'major_cov'] = df.loc[evidence_row_index, 'major_cov']\n",
    "                df.loc[entry_row_index, 'minor_cov'] = df.loc[evidence_row_index, 'minor_cov']\n",
    "                df.loc[entry_row_index, 'total_cov'] = df.loc[evidence_row_index, 'total_cov']\n",
    "                df.loc[entry_row_index, 'ra_cov'] = df.loc[evidence_row_index, 'ra_cov']\n",
    "                df.loc[entry_row_index, 'jc_cov'] = df.loc[evidence_row_index, 'jc_cov']\n",
    "                df.loc[entry_row_index, 'mc_cov'] = df.loc[evidence_row_index, 'mc_cov']\n",
    "            count += 1\n",
    "    return df\n",
    "\n",
    "def get_mutations_df(gdfile, culture_name, generation=np.nan, cov=False):\n",
    "    '''\n",
    "    Adds new columns for culture name, and sampled generation.\n",
    "    Returns df of only mutation entry rows, with added evidence column.\n",
    "    '''\n",
    "    if cov == True:\n",
    "        gd_subset = subset_gd_to_df(gdfile, cov=True)    \n",
    "        gd_subset_df = add_evidence_to_mutation_rows(gd_subset, cov=True)\n",
    "    else:\n",
    "        gd_subset = subset_gd_to_df(gdfile)    \n",
    "        gd_subset_df = add_evidence_to_mutation_rows(gd_subset)\n",
    "    gd_subset_df.insert(0, 'culture', culture_name)\n",
    "    gd_subset_df.insert(1, 'generation', generation)\n",
    "    mutations_df = select_mutation_rows(gd_subset_df).copy()\n",
    "    return mutations_df\n",
    "\n",
    "def get_population_subset_frequencies_to_plot(population_df, subset_df):\n",
    "    '''\n",
    "    Merge two mutation dfs to plot frequencies of mutations in common.\n",
    "    '''\n",
    "    merged = pd.merge(population_df, subset_df, on=('genome_id', 'position'))\n",
    "    population_frequencies_df = merged.drop(['culture_y', 'generation_y', 'entry_type_y', 'entry_id_y', 'evidence_id_y', \n",
    "                                          'mutation_detail_y', 'frequency_y', 'gene_product_y', 'reject_y', \n",
    "                                          'prediction_y', 'polymorphism_frequency_y'], axis=1)\n",
    "    reshaped = population_frequencies_df.pivot_table(index=['genome_id', 'position', 'entry_type_x', 'mutation_detail_x'], \n",
    "                                                     columns='generation_x', \n",
    "                                                     values='polymorphism_frequency_x').fillna(0).T\n",
    "    return reshaped\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = get_mutations_df('/Users/ymseah/Repositories/dvh_mms2/depth_mle/d2/annotated.gd', 'D2', 0, cov=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2.to_csv('/Users/ymseah/Repositories/dvh_mms2/depth_mle/d2/d2mut_cov.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda]",
   "language": "python",
   "name": "conda-env-anaconda-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
